// @microsoft/sp-adaptive-card-extension-base/lib-commonjs/ac-template/util/actionButtonHelper.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.updateSubmitActionButtonType = exports.createAction = exports.parseButtonsToAction = exports.appendActions = void 0;
var tslib_1 = require("tslib");
var sp_core_library_1 = require("@microsoft/sp-core-library");
var Strings_resx_1 = tslib_1.__importDefault(require("../../loc/Strings.resx"));
var Killswitches_1 = require("../../common/Killswitches");
// Maps card size to max number of actions that can be appended.
var SIZE_TO_ACTION_NUM_MAP = {
    Medium: 1,
    Large: 2
};
function appendActions(template, buttons, size, templateType, instanceId) {
    if (!(0, Killswitches_1.isRefactorParseCardButtonsToAction)()) {
        var actions = buttons && parseButtonsToAction(buttons, templateType, size, instanceId);
        if (actions) {
            template.actions = actions;
        }
    }
    else {
        var actions = [];
        var limit = Math.min(SIZE_TO_ACTION_NUM_MAP[size], (buttons === null || buttons === void 0 ? void 0 : buttons.length) || 0);
        for (var i = 0; i < limit; i++) {
            var cardButton = buttons[i];
            var actionId = cardButton.id
                ? cardButton.id
                : !(0, Killswitches_1.isGenerateUniqueButtonIdKSActivated)()
                    ? "ac-button-".concat(instanceId, "-").concat(i)
                    : "ac-button-".concat(i);
            actions.push(createAction(cardButton.action, actionId, cardButton.title, cardButton.style));
        }
        template.actions = actions;
    }
}
exports.appendActions = appendActions;
function parseButtonsToAction(buttons, templateType, cardSize, instanceId) {
    var filteredButtons = filterButtons(buttons, templateType, cardSize);
    if (!filteredButtons) {
        return undefined;
    }
    return filteredButtons.map(function (button, index) {
        var actionId = button.id
            ? button.id
            : !(0, Killswitches_1.isGenerateUniqueButtonIdKSActivated)()
                ? "ac-button-".concat(instanceId, "-").concat(index)
                : "ac-button-".concat(index);
        return createAction(button.action, actionId, button.title, button.style);
    });
}
exports.parseButtonsToAction = parseButtonsToAction;
function filterButtons(buttons, templateType, cardSize) {
    if ((templateType === 'Image' && cardSize === 'Medium') || buttons === undefined) {
        return undefined;
    }
    var filteredButtons = [];
    var limit = Math.min(SIZE_TO_ACTION_NUM_MAP[cardSize], buttons.length);
    for (var i = 0; i < limit; i++) {
        if ((0, Killswitches_1.isSupportActionExecuteKSActivated)() && buttons[i].action.type === 'Execute') {
            continue;
        }
        filteredButtons.push(buttons[i]);
    }
    return filteredButtons;
}
/**
 * @internal
 */
function createAction(buttonType, actionId, title, style, target) {
    var cardAction;
    //
    // with strictNullCheck set to false a developer can pass undefined as a buttonType
    // we want to throw an error in this case to allow a developer to fix the issue
    //
    if (!buttonType && !(0, Killswitches_1.isThrowErrorWhenButtonActionIsUndefined)()) {
        throw new Error(sp_core_library_1.Text.format(Strings_resx_1.default.ActionUndefined, actionId));
    }
    if (buttonType.type === 'Submit') {
        updateSubmitActionButtonType(buttonType);
        cardAction = {
            id: actionId,
            type: 'Action.Submit',
            style: style,
            title: title,
            data: buttonType.parameters || undefined
        };
    }
    else if (buttonType.type === 'Execute') {
        cardAction = {
            id: actionId,
            type: 'Action.Execute',
            style: style,
            title: title,
            verb: buttonType.verb,
            data: buttonType.parameters || undefined
        };
    }
    else if (buttonType.type === 'ExternalLink') {
        var url = buttonType.parameters.target;
        if (!url || url === '') {
            throw new Error(sp_core_library_1.Text.format(Strings_resx_1.default.TargetUrlIsUndefinedOrEmpty, actionId));
        }
        cardAction = {
            id: actionId,
            type: 'Action.OpenUrl',
            style: style,
            title: title,
            url: url
        };
    }
    else if (buttonType.type === 'QuickView') {
        var data = {
            _internalQuickViewId: buttonType.parameters.view,
            _target: target
        };
        cardAction = {
            id: actionId,
            type: 'Action.Submit',
            style: style,
            title: title,
            data: data
        };
    }
    else if (buttonType.type === 'VivaAction.SelectMedia') {
        var parameters = buttonType.parameters;
        cardAction = {
            id: actionId,
            type: 'VivaAction.SelectMedia',
            style: style,
            title: title,
            parameters: parameters
        };
    }
    else if (buttonType.type === 'VivaAction.GetLocation') {
        var parameters = buttonType.parameters;
        cardAction = {
            id: actionId,
            type: 'VivaAction.GetLocation',
            style: style,
            title: title,
            parameters: parameters
        };
    }
    else if (buttonType.type === 'VivaAction.ShowLocation') {
        var parameters = buttonType.parameters;
        cardAction = {
            id: actionId,
            type: 'VivaAction.ShowLocation',
            style: style,
            title: title,
            parameters: parameters
        };
    }
    return cardAction;
}
exports.createAction = createAction;
/**
 * Inserts confirmation dialog into action.parameters as __MSConfirmationDialog__
 * @param action - ISubmitCardAction
 */
function updateSubmitActionButtonType(action) {
    if (action.confirmationDialog) {
        if (!action.parameters) {
            action.parameters = {};
        }
        action.parameters.__MSConfirmationDialog__ = action.confirmationDialog;
    }
}
exports.updateSubmitActionButtonType = updateSubmitActionButtonType;
