// @microsoft/sp-http-base/lib-commonjs/oauthTokenProvider/AadFlights.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.isSafariAuthPatchEnbled = exports.isOBOForAllTeamsScenariosEnabled = exports.isUseMsalCurrentChunkEnabled = exports.isAadTokenProviderPopupFlowV1Enabled = exports.isPoPTokenFlightEnabled = exports.isMsalBrowserFirstPartyBuildFlightEnabled = exports.isMsalBrowserFlightEnabled = void 0;
var sp_core_library_1 = require("@microsoft/sp-core-library");
function isMsalBrowserFlightEnabled() {
    return sp_core_library_1._SPFlight.isEnabled(60346);
}
exports.isMsalBrowserFlightEnabled = isMsalBrowserFlightEnabled;
function isMsalBrowserFirstPartyBuildFlightEnabled() {
    return sp_core_library_1._SPFlight.isEnabled(1365);
}
exports.isMsalBrowserFirstPartyBuildFlightEnabled = isMsalBrowserFirstPartyBuildFlightEnabled;
function isPoPTokenFlightEnabled() {
    return sp_core_library_1._SPFlight.isEnabled(1512);
}
exports.isPoPTokenFlightEnabled = isPoPTokenFlightEnabled;
function isAadTokenProviderPopupFlowV1Enabled() {
    return sp_core_library_1._SPFlight.isEnabled(1791);
}
exports.isAadTokenProviderPopupFlowV1Enabled = isAadTokenProviderPopupFlowV1Enabled;
function isUseMsalCurrentChunkEnabled() {
    return sp_core_library_1._SPFlight.isEnabled(60835);
}
exports.isUseMsalCurrentChunkEnabled = isUseMsalCurrentChunkEnabled;
/**
 * APEXUseOboForAllTeamsScenarios
 */
function isOBOForAllTeamsScenariosEnabled() {
    return sp_core_library_1._SPFlight.isEnabled(60244);
}
exports.isOBOForAllTeamsScenariosEnabled = isOBOForAllTeamsScenariosEnabled;
/**
 * SppplatSafariAuthPatch
 */
function isSafariAuthPatchEnbled() {
    return sp_core_library_1._SPFlight.isEnabled(687);
}
exports.isSafariAuthPatchEnbled = isSafariAuthPatchEnbled;
