// @microsoft/sp-http-base/lib-commonjs/oauthTokenProvider/AadKillSwitches.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.isMsalImplicitV1TokenByPopupKSActivated = exports.isLogAfterFullPageRedirectKSActivated = exports.useOBOInWebForSecureBrokerKSActivated = exports.isTeamsBrowserHostedKSActive = exports.isPreventInfiniteRefreshViaReadingFromSessionStorageKSActivated = exports.isMsalUseSeparateStorageKeysKsActivated = exports.isMsalBrowserRetryAndSsoSilentImprovementsKsActivated = exports.isMsalTwoMinuteMaxRedirectKsActivated = exports.isAadTokenErrorLogKsActivated = exports.isExtendedErrorObjectActivated = exports.isSkipKnownErroneousResourcesActivated = void 0;
var sp_core_library_1 = require("@microsoft/sp-core-library");
function isSkipKnownErroneousResourcesActivated() {
    return sp_core_library_1._SPKillSwitch.isActivated('dd3dba4e-4baa-4bd3-95bf-64ffae9b7b51'
    /* '01/04/22', 'Skip making a request for resources that have already failed' */
    );
}
exports.isSkipKnownErroneousResourcesActivated = isSkipKnownErroneousResourcesActivated;
function isExtendedErrorObjectActivated() {
    return sp_core_library_1._SPKillSwitch.isActivated('a4a3d2f4-ad0e-4684-adee-a5ceec193558'
    /* '03/08/22', 'Extends error object to allow for more properties to be filled' */
    );
}
exports.isExtendedErrorObjectActivated = isExtendedErrorObjectActivated;
function isAadTokenErrorLogKsActivated() {
    return sp_core_library_1._SPKillSwitch.isActivated('e212149a-997e-4aa9-aa89-8be796153d3e'
    /* '05/27/2022',
        'Log actual acquire token monitor error'*/
    );
}
exports.isAadTokenErrorLogKsActivated = isAadTokenErrorLogKsActivated;
function isMsalTwoMinuteMaxRedirectKsActivated() {
    return sp_core_library_1._SPKillSwitch.isActivated('ca6e64a5-3cde-4350-8a45-a21de378cc60'
    /* '07/14/22', 'Enable MSAL to hit max redirect limit in 2 minutes instead of 1' */
    );
}
exports.isMsalTwoMinuteMaxRedirectKsActivated = isMsalTwoMinuteMaxRedirectKsActivated;
function isMsalBrowserRetryAndSsoSilentImprovementsKsActivated() {
    return sp_core_library_1._SPKillSwitch.isActivated('e0bd3586-89a3-466b-8ee0-db1f9457fe99'
    /* '07/14/22', 'Stop MSAL Browser from retrying endpoint resolution failures + increase timeout time' */
    );
}
exports.isMsalBrowserRetryAndSsoSilentImprovementsKsActivated = isMsalBrowserRetryAndSsoSilentImprovementsKsActivated;
function isMsalUseSeparateStorageKeysKsActivated() {
    return sp_core_library_1._SPKillSwitch.isActivated('1a423106-78a1-4238-ba44-dd5f7f4138df'
    /* '08/18/22', 'Separate the storage parameters used for redirect flows by v1/v2' */
    );
}
exports.isMsalUseSeparateStorageKeysKsActivated = isMsalUseSeparateStorageKeysKsActivated;
function isPreventInfiniteRefreshViaReadingFromSessionStorageKSActivated() {
    return sp_core_library_1._SPKillSwitch.isActivated('fed6018f-7453-47fc-aecb-2d9505b62bb4'
    /* '03/02/23', 'Add session storage check to prevent infinite refresh' */
    );
}
exports.isPreventInfiniteRefreshViaReadingFromSessionStorageKSActivated = isPreventInfiniteRefreshViaReadingFromSessionStorageKSActivated;
function isTeamsBrowserHostedKSActive() {
    return sp_core_library_1._SPKillSwitch.isActivated('733da2bd-f34d-4979-8c7b-df7342291267'
    /* '06/22/2023', 'Set teams hosted container' */
    );
}
exports.isTeamsBrowserHostedKSActive = isTeamsBrowserHostedKSActive;
function useOBOInWebForSecureBrokerKSActivated() {
    return sp_core_library_1._SPKillSwitch.isActivated('043feae9-5a56-40ff-b175-6f7f813507ff' /* '10/26/22', 'use OBO in web for secure broker' */);
}
exports.useOBOInWebForSecureBrokerKSActivated = useOBOInWebForSecureBrokerKSActivated;
function isLogAfterFullPageRedirectKSActivated() {
    return sp_core_library_1._SPKillSwitch.isActivated('5a350491-8d32-46a3-a4c4-fdce19a5f73e' /* '08/17/23', 'log after full page redirect' */);
}
exports.isLogAfterFullPageRedirectKSActivated = isLogAfterFullPageRedirectKSActivated;
function isMsalImplicitV1TokenByPopupKSActivated() {
    return sp_core_library_1._SPKillSwitch.isActivated('e6246a29-f6af-41de-88b7-3f023102181b' /* '08/14/23', 'Enable acquiring a token by popup' */);
}
exports.isMsalImplicitV1TokenByPopupKSActivated = isMsalImplicitV1TokenByPopupKSActivated;
