// @microsoft/spfx-heft-plugins/lib/plugins/PluginNames.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.DeployAzureStoragePlugin = exports.ThirdPartyExternalsPlugin = exports.PackageSolutionPlugin = exports.CustomizeWebpackConfigPlugin = exports.SpfxWebpackServePlugin = exports.SpfxWebpackConfigurationPlugin = exports.SpfxHeftExtensionsPlugin = exports.SpfxHeftPlugin = exports.DevDeployPlugin = void 0;
/**
 * @public
 */
exports.DevDeployPlugin = 'spfx-dev-deploy-plugin';
/**
 * @public
 */
exports.SpfxHeftPlugin = 'spfx-heft-plugin';
/**
 * @public
 */
exports.SpfxHeftExtensionsPlugin = 'spfx-heft-extensions-plugin';
/**
 * @public
 */
exports.SpfxWebpackConfigurationPlugin = 'spfx-webpack-configuration-plugin';
/**
 * @public
 */
exports.SpfxWebpackServePlugin = 'spfx-webpack-serve-plugin';
/**
 * @public
 */
exports.CustomizeWebpackConfigPlugin = 'customize-spfx-webpack-configuration-plugin';
/**
 * @public
 */
exports.PackageSolutionPlugin = 'package-solution-plugin';
/**
 * @public
 */
exports.ThirdPartyExternalsPlugin = 'third-party-externals-plugin';
/**
 * @public
 */
exports.DeployAzureStoragePlugin = 'deploy-azure-storage-plugin';
