// @microsoft/sp-loader/lib-commonjs/utilities/componentConstants.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.spApplicationBaseName = exports.microsoftGraphClientComponentId = exports.tslibComponentId = exports.classicPageAppComponentId = exports.listViewHostComponentId = exports.spApplicationBaseComponentId = exports.reactDomProfilingComponentId = exports.reactDomDebugComponentId = exports.reactDomComponentId = exports.reactDebugComponentId = exports.reactComponentId = void 0;
// IDs
exports.reactComponentId = '0d910c1c-13b9-4e1c-9aa4-b008c5e42d7d';
exports.reactDebugComponentId = '4109cd8f-dfb9-4fd2-99b6-22cf09dbb709';
exports.reactDomComponentId = 'aa0a46ec-1505-43cd-a44a-93f3a5aa460a';
exports.reactDomDebugComponentId = '6c91b906-94e9-4c81-a65f-265b97c019b6';
exports.reactDomProfilingComponentId = 'd0619256-3350-4eb2-95df-31fbc25c4451';
exports.spApplicationBaseComponentId = '4df9bb86-ab0a-4aab-ab5f-48bf167048fb';
exports.listViewHostComponentId = 'b1ab4aaa-f779-405c-8683-d3a750b5d18d';
exports.classicPageAppComponentId = 'eb4b666b-5c29-4dad-9a99-23613f21a2b7';
exports.tslibComponentId = '01c4df03-e775-48cb-aa14-171ee5199a15';
exports.microsoftGraphClientComponentId = '32b9363b-09ff-4959-929a-cab2663fb223';
// Names
exports.spApplicationBaseName = '@microsoft/sp-application-base';
