// @microsoft/sp-loader/lib-commonjs/utilities/telemetryConstants.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.spRequireJsComponentLoader = exports.manifestStoreLogSource = exports.componentStoreLogSource = exports.resolveAddressLogSource = exports.loadScriptLogSource = exports.loadComponentLogSource = exports.startApplicationLogSource = exports.spStarterLogSource = exports.noSecondaryCdnUrl = exports.usingSecondaryCdnEventName = exports.loadPathDependenciesErrorTagName = exports.loadComponentDependenciesErrorTagName = exports.configureSystemJsErrorTagName = exports.emptyComponentErrorTagName = exports.loadComponentWithFallbackSystemQosScenarioName = exports.loadComponentWithRequireQosScenarioName = exports.loadScriptWithRequireQosScenarioName = exports.loadPathDependencyQosScenarioName = exports.loadComponentImplQosScenarioName = exports.loadComponentQosScenarioName = exports.loadScriptQosScenarioName = void 0;
// ISPComponentLoader.loadComponent
exports.loadScriptQosScenarioName = 'SPComponentLoader.loadScript';
exports.loadComponentQosScenarioName = 'SPComponentLoader.loadComponent';
exports.loadComponentImplQosScenarioName = 'SPComponentLoader._loadComponentImpl';
exports.loadPathDependencyQosScenarioName = 'SPComponentLoader.loadPathDependency';
exports.loadScriptWithRequireQosScenarioName = 'SPComponentLoader.loadScriptWithRequire';
exports.loadComponentWithRequireQosScenarioName = 'SPComponentLoader.loadComponentWithRequire';
exports.loadComponentWithFallbackSystemQosScenarioName = 'SPComponentLoader.loadComponentWithFallbackSystem';
exports.emptyComponentErrorTagName = 'EmptyComponent';
exports.configureSystemJsErrorTagName = 'ConfigureSystemJs';
exports.loadComponentDependenciesErrorTagName = 'LoadComponentDependencies';
exports.loadPathDependenciesErrorTagName = 'LoadPathDependencies';
exports.usingSecondaryCdnEventName = 'UsingSecondaryCdn';
exports.noSecondaryCdnUrl = 'NoSecondaryCdnUrl';
// Log sources
exports.spStarterLogSource = { id: 'SPStarter.start' };
exports.startApplicationLogSource = { id: 'SPComponentLoader.startApplication' };
exports.loadComponentLogSource = { id: 'SPComponentLoader.loadComponent' };
exports.loadScriptLogSource = { id: 'SPComponentLoader.loadScript' };
exports.resolveAddressLogSource = { id: 'resolveAddress' };
exports.componentStoreLogSource = { id: 'ComponentStore' };
exports.manifestStoreLogSource = { id: 'ManifestStore' };
exports.spRequireJsComponentLoader = { id: 'SPRequireJsComponentLoader' };
