// @microsoft/sp-component-base/lib-commonjs/index.js
"use strict";
/**
 * SharePoint Framework support for building client-side components
 *
 * @remarks
 * This package supports development of client-side components (e.g. web parts,
 * client-side extensions, and applications) built using the SharePoint Framework.
 * It defines the base classes and interfaces for functionality that is
 * common across all client-side component types.
 *
 * @packagedocumentation
 */
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports._CacheMissReason = exports._WebPartLoadDataCollector = exports._ensureIsolatedDomainUrl = exports._ShimmerFactory = exports.ThemeProvider = exports.ThemeChangedEventArgs = exports.ServerProcessedDataType = exports._PropertyMetadataProcessor = exports._LegacyThirdPartyFabricCoreLoader = exports.DynamicProperty = exports.DynamicDataSourceManager = exports.DynamicDataProvider = exports.BaseComponentContext = exports.BaseComponent = void 0;
var BaseComponent_1 = require("./BaseComponent");
Object.defineProperty(exports, "BaseComponent", { enumerable: true, get: function () { return __importDefault(BaseComponent_1).default; } });
var BaseComponentContext_1 = require("./BaseComponentContext");
Object.defineProperty(exports, "BaseComponentContext", { enumerable: true, get: function () { return __importDefault(BaseComponentContext_1).default; } });
var DynamicDataProvider_1 = require("./DynamicDataProvider");
Object.defineProperty(exports, "DynamicDataProvider", { enumerable: true, get: function () { return __importDefault(DynamicDataProvider_1).default; } });
var DynamicDataSourceManager_1 = require("./DynamicDataSourceManager");
Object.defineProperty(exports, "DynamicDataSourceManager", { enumerable: true, get: function () { return __importDefault(DynamicDataSourceManager_1).default; } });
var DynamicProperty_1 = require("./DynamicProperty");
Object.defineProperty(exports, "DynamicProperty", { enumerable: true, get: function () { return __importDefault(DynamicProperty_1).default; } });
var LegacyThirdPartyFabricCoreLoader_1 = require("./LegacyThirdPartyFabricCoreLoader");
Object.defineProperty(exports, "_LegacyThirdPartyFabricCoreLoader", { enumerable: true, get: function () { return __importDefault(LegacyThirdPartyFabricCoreLoader_1).default; } });
var PropertyMetadataProcessor_1 = require("./PropertyMetadataProcessor");
Object.defineProperty(exports, "_PropertyMetadataProcessor", { enumerable: true, get: function () { return PropertyMetadataProcessor_1.PropertyMetadataProcessor; } });
var ISerializedServerProcessedData_1 = require("./ISerializedServerProcessedData");
Object.defineProperty(exports, "ServerProcessedDataType", { enumerable: true, get: function () { return ISerializedServerProcessedData_1.ServerProcessedDataType; } });
var ThemeProvider_1 = require("./themeProvider/ThemeProvider");
Object.defineProperty(exports, "ThemeChangedEventArgs", { enumerable: true, get: function () { return ThemeProvider_1.ThemeChangedEventArgs; } });
Object.defineProperty(exports, "ThemeProvider", { enumerable: true, get: function () { return ThemeProvider_1.ThemeProvider; } });
// ShimmerFactory
var ShimmerFactory_1 = require("./ShimmerFactory");
Object.defineProperty(exports, "_ShimmerFactory", { enumerable: true, get: function () { return __importDefault(ShimmerFactory_1).default; } });
var EnsureIsolatedDomainUrl_1 = require("./EnsureIsolatedDomainUrl");
Object.defineProperty(exports, "_ensureIsolatedDomainUrl", { enumerable: true, get: function () { return EnsureIsolatedDomainUrl_1.ensureIsolatedDomainUrl; } });
var WebPartLoadDataCollector_1 = require("./dataCollector/WebPartLoadDataCollector");
Object.defineProperty(exports, "_WebPartLoadDataCollector", { enumerable: true, get: function () { return WebPartLoadDataCollector_1.WebPartLoadDataCollector; } });
var IWebPartLoadData_1 = require("./dataCollector/IWebPartLoadData");
Object.defineProperty(exports, "_CacheMissReason", { enumerable: true, get: function () { return IWebPartLoadData_1.CacheMissReason; } });
