// @microsoft/sp-adaptive-card-extension-base/lib-commonjs/views/card/BaseImageCardView.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.BaseImageCardView = void 0;
var tslib_1 = require("tslib");
var imageL1Template_1 = require("../../ac-template/imageL1Template");
var imageL1TemplateSelectAction_1 = require("../../ac-template/imageL1TemplateSelectAction");
var BaseTemplateCardView_1 = require("./BaseTemplateCardView");
var GenericTemplates_1 = require("./parameters/GenericTemplates");
/**
 * Base class for an Image Card view.
 *
 * @remarks The Image Card view displays the following:
 *  - Card title
 *  - Card icon
 *  - One image
 *  - One primary text field
 *  - Zero buttons in the `Medium` Card size, up to two buttons in `Large` Card size
 *
 * @public
 */
var BaseImageCardView = /** @class */ (function (_super) {
    tslib_1.__extends(BaseImageCardView, _super);
    function BaseImageCardView() {
        var _this = _super !== null && _super.apply(this, arguments) || this;
        /**
         * @internal
         */
        _this._templateType = 'Image';
        return _this;
    }
    /**
     * @internal
     */
    BaseImageCardView.prototype._getCardViewParameters = function () {
        var data = this.data;
        var parameters = (0, GenericTemplates_1.ImageCardView)({
            image: {
                url: data.imageUrl,
                altText: data.iconAltText,
                _urlFallback: data._imageUrlFallback
            },
            cardBar: this._getCardBarConfiguration(),
            header: {
                componentName: 'text',
                text: data.primaryText
            },
            footer: this._getGenericCardViewFooterConfiguration()
        });
        // updating the cardViewType to track "old" templates
        parameters.cardViewType = 'image-template';
        return parameters;
    };
    Object.defineProperty(BaseImageCardView.prototype, "template", {
        /**
         * Returns the Adaptive Card template for the {@link BaseImageCardView}.
         *
         * @public
         * @sealed
         *
         * @deprecated template property is deprecated for card views.
         */
        get: function () {
            return this._getTemplate(imageL1Template_1.imageL1Template, imageL1TemplateSelectAction_1.imageL1TemplateSelectAction);
        },
        enumerable: false,
        configurable: true
    });
    return BaseImageCardView;
}(BaseTemplateCardView_1.BaseTemplateCardView));
exports.BaseImageCardView = BaseImageCardView;
