// @microsoft/sp-adaptive-card-extension-base/lib-commonjs/views/card/BasePrimaryTextCardView.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.BasePrimaryTextCardView = void 0;
var tslib_1 = require("tslib");
var primaryTextL1Template_1 = require("../../ac-template/primaryTextL1Template");
var primaryTextL1TemplateSelectAction_1 = require("../../ac-template/primaryTextL1TemplateSelectAction");
var BaseTemplateCardView_1 = require("./BaseTemplateCardView");
var GenericTemplates_1 = require("./parameters/GenericTemplates");
/**
 * Base class for a Primary Text Card view.
 *
 * @remarks The Primary Text Card view displays the following:
 *  - Card Title
 *  - Card Icon
 *  - One primary text field
 *  - One description text field
 *  - One buttons in the `Medium` Card size, up to two buttons in `Large` Card size
 *
 *
 * @public
 */
var BasePrimaryTextCardView = /** @class */ (function (_super) {
    tslib_1.__extends(BasePrimaryTextCardView, _super);
    function BasePrimaryTextCardView() {
        var _this = _super !== null && _super.apply(this, arguments) || this;
        /**
         * @internal
         */
        _this._templateType = 'PrimaryText';
        return _this;
    }
    /**
     * @internal
     */
    BasePrimaryTextCardView.prototype._getCardViewParameters = function () {
        var data = this.data;
        var parameters = (0, GenericTemplates_1.PrimaryTextCardView)({
            cardBar: this._getCardBarConfiguration(),
            header: {
                componentName: 'text',
                text: data.primaryText
            },
            body: {
                componentName: 'text',
                text: data.description
            },
            footer: this._getGenericCardViewFooterConfiguration()
        });
        // updating the cardViewType to track "old" templates
        parameters.cardViewType = 'primaryText-template';
        return parameters;
    };
    Object.defineProperty(BasePrimaryTextCardView.prototype, "template", {
        /**
         * Returns the Adaptive Card template for the {@link BasePrimaryTextCardView}.
         *
         * @public
         * @sealed
         * @deprecated template property is deprecated for card views.
         */
        get: function () {
            return this._getTemplate(primaryTextL1Template_1.primaryTextL1Template, primaryTextL1TemplateSelectAction_1.primaryTextL1TemplateSelectAction);
        },
        enumerable: false,
        configurable: true
    });
    return BasePrimaryTextCardView;
}(BaseTemplateCardView_1.BaseTemplateCardView));
exports.BasePrimaryTextCardView = BasePrimaryTextCardView;
