// @microsoft/sp-adaptive-card-extension-base/lib-commonjs/views/card/BaseBasicCardView.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.BaseBasicCardView = void 0;
var tslib_1 = require("tslib");
var basicL1Template_1 = require("../../ac-template/basicL1Template");
var basicL1TemplateSelectAction_1 = require("../../ac-template/basicL1TemplateSelectAction");
var BaseTemplateCardView_1 = require("./BaseTemplateCardView");
var GenericTemplates_1 = require("./parameters/GenericTemplates");
/**
 * Base class for a Basic Card view.
 *
 * @remarks The basic Card view displays the following:
 *  - Card title
 *  - Card icon
 *  - One primary text field
 *  - One button in the `Medium` Card size, two buttons in the `Large` Card size
 *
 * @public
 */
var BaseBasicCardView = /** @class */ (function (_super) {
    tslib_1.__extends(BaseBasicCardView, _super);
    function BaseBasicCardView() {
        var _this = _super !== null && _super.apply(this, arguments) || this;
        /**
         * @internal
         */
        _this._templateType = 'Basic';
        return _this;
    }
    /**
     * @internal
     */
    BaseBasicCardView.prototype._getCardViewParameters = function () {
        var data = this.data;
        var parameters = (0, GenericTemplates_1.BasicCardView)({
            cardBar: this._getCardBarConfiguration(),
            header: {
                componentName: 'text',
                text: data.primaryText
            },
            footer: this._getGenericCardViewFooterConfiguration()
        });
        // updating the cardViewType to track "old" templates
        parameters.cardViewType = 'basic-template';
        return parameters;
    };
    Object.defineProperty(BaseBasicCardView.prototype, "template", {
        /**
         * Returns the Adaptive Card template for the {@link BaseBasicCardView}.
         *
         * @public
         * @sealed
         * @deprecated template property is deprecated for card views.
         */
        get: function () {
            return this._getTemplate(basicL1Template_1.basicL1Template, basicL1TemplateSelectAction_1.basicL1TemplateSelectAction);
        },
        enumerable: false,
        configurable: true
    });
    return BaseBasicCardView;
}(BaseTemplateCardView_1.BaseTemplateCardView));
exports.BaseBasicCardView = BaseBasicCardView;
