// @microsoft/sp-adaptive-card-extension-base/lib-commonjs/common/Killswitches.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.isPrepareHTMLQVForMessagingKSActivated = exports.isProvideCardViewIdKSActivated = exports.isGenerateComponentIdsKSActivated = exports.isCatchNullBotResponseDataKSActivated = exports.isAddParentWindowOriginQSPToIsolatedPageKSActivated = exports.isRemoveFunctionsForMessagingKSActivated = exports.isMovePropertiesFromDataToCardPropertiesKSActivated = exports.isEnableHTMLQuickViewKSActivated = exports.isFixIframeNameKSActive = exports.isThrowErrorWhenButtonActionIsUndefined = exports.isMoveBaseEngagementDataInSeparateFunctionKSActivated = exports.isAddVetoForACEKSActivated = exports.isLockUpdateKSActivated = exports.isSupportActionExecuteKSActivated = exports.isLogEngagementExtraDataKSActivated = exports.isAddAppHostTypeToCardActionKSActivated = exports.isRemovingViewNavigatorInstantiationFromConstructorKSActivated = exports.isIconAltTextNotOverriddenKSActivated = exports.isFixOnBeforeActionExternalLinkTelemetryKSActivated = exports.isDisableVivaConnectionsAnalyticsEventsKSActivated = exports.isSurfaceOnBeforeActionKSActivated = exports.isACECardViewReactEnabledOnSecureBrokerKSActivated = exports.isSafeCheckUndefinedCacheProviderKSActivated = exports.isThrowErrorWhenRenderArgsAreUndefined = exports.isGenerateUniqueButtonIdKSActivated = exports.isRemoveIsolatedDomainOnDisposeKSActivated = exports.isAddDefaultImageAltTextKSActivated = exports.isLogACECacheValidityKSActivated = exports.isReduceRenderTimeoutKSActivated = exports.isExcludeVivaHomeFromMySiteCacheKSActivated = exports.isRefactorParseCardButtonsToAction = exports.isTargetBotDriveAceAspxKSActivated = exports.isIsolatedAdaptiveCardExtensionWebPropertyPaneWorkAroundKSActivated = void 0;
var sp_core_library_1 = require("@microsoft/sp-core-library");
function isIsolatedAdaptiveCardExtensionWebPropertyPaneWorkAroundKSActivated() {
    return sp_core_library_1._SPKillSwitch.isActivated('b8c50ded-d048-4eb9-bc41-21190d9d3bbb'
    /* '03/17/2022', 'Work around to enable property pane for isolated ACE web' */
    );
}
exports.isIsolatedAdaptiveCardExtensionWebPropertyPaneWorkAroundKSActivated = isIsolatedAdaptiveCardExtensionWebPropertyPaneWorkAroundKSActivated;
function isTargetBotDriveAceAspxKSActivated() {
    return sp_core_library_1._SPKillSwitch.isActivated('e8daf7db-132e-4225-93a4-70cc5fc06bb9'
    /* '04/14/22', 'Target botDriveAce.aspx for Bot Drive ACEs.' */
    );
}
exports.isTargetBotDriveAceAspxKSActivated = isTargetBotDriveAceAspxKSActivated;
function isRefactorParseCardButtonsToAction() {
    return sp_core_library_1._SPKillSwitch.isActivated('db2c6f89-2750-4b0c-87f4-fd4b28af598e'
    /* '10/09/2022', 'Hoist reusable methods to parse and append buttons to action to keep code base DRY.' */
    );
}
exports.isRefactorParseCardButtonsToAction = isRefactorParseCardButtonsToAction;
function isExcludeVivaHomeFromMySiteCacheKSActivated() {
    return sp_core_library_1._SPKillSwitch.isActivated('5c806b76-07fd-4ea8-afce-10c676a1739e'
    /* '10/11/2022', 'Exclude VivaHome from using ACE MySiteCache.' */
    );
}
exports.isExcludeVivaHomeFromMySiteCacheKSActivated = isExcludeVivaHomeFromMySiteCacheKSActivated;
function isReduceRenderTimeoutKSActivated() {
    return sp_core_library_1._SPKillSwitch.isActivated('8a4feb2f-ae4e-4174-9be1-0eb0b1371665'
    /* '10/17/2022', 'Reduce default render timeout of ACEs to 10s.' */
    );
}
exports.isReduceRenderTimeoutKSActivated = isReduceRenderTimeoutKSActivated;
function isLogACECacheValidityKSActivated() {
    return sp_core_library_1._SPKillSwitch.isActivated('63a9cba3-4d6e-44a7-bdfb-3464aeb6a4a4'
    /* '11/02/2022', 'Logs if ace cache is valid of has invalidated due to time out.' */
    );
}
exports.isLogACECacheValidityKSActivated = isLogACECacheValidityKSActivated;
function isAddDefaultImageAltTextKSActivated() {
    return sp_core_library_1._SPKillSwitch.isActivated('63a9cba3-4d6e-44a7-bdfb-3464aeb6a4a4'
    /* '11/22/2022', 'Adds default alt text for icons and images rendered in the card view.' */
    );
}
exports.isAddDefaultImageAltTextKSActivated = isAddDefaultImageAltTextKSActivated;
function isRemoveIsolatedDomainOnDisposeKSActivated() {
    return sp_core_library_1._SPKillSwitch.isActivated('080770ac-5b8a-4f68-97b3-abf37230c2ce'
    /* '11/23/2022', 'When disposing of an isolated ACE WP, dispose of the isolated frame in the isolated service.' */
    );
}
exports.isRemoveIsolatedDomainOnDisposeKSActivated = isRemoveIsolatedDomainOnDisposeKSActivated;
function isGenerateUniqueButtonIdKSActivated() {
    return sp_core_library_1._SPKillSwitch.isActivated('aebbeddc-284c-470d-8bd2-98a4d200cb86'
    /* '11/28/2022', 'Use instance id as part of button id to make it unique.' */
    );
}
exports.isGenerateUniqueButtonIdKSActivated = isGenerateUniqueButtonIdKSActivated;
function isThrowErrorWhenRenderArgsAreUndefined() {
    return sp_core_library_1._SPKillSwitch.isActivated('04fc0392-7564-42c1-9bb4-37b2844d13c6'
    /* '04fc0392-7564-42c1-9bb4-37b2844d13c6', 'Throw error when render arguments for the ACE are undefined.' */
    );
}
exports.isThrowErrorWhenRenderArgsAreUndefined = isThrowErrorWhenRenderArgsAreUndefined;
function isSafeCheckUndefinedCacheProviderKSActivated() {
    return sp_core_library_1._SPKillSwitch.isActivated('c761d258-674d-4c2c-91ca-37cd43dd78a5'
    /* '12/19/2022', 'In the case ACE MySiteCache cacheProvider is undefined avoid using caching feature.' */
    );
}
exports.isSafeCheckUndefinedCacheProviderKSActivated = isSafeCheckUndefinedCacheProviderKSActivated;
function isACECardViewReactEnabledOnSecureBrokerKSActivated() {
    return sp_core_library_1._SPKillSwitch.isActivated('e0749332-8697-4811-b063-50b25956b1a3'
    /* '01/23/2023', 'Enable ACE card view React on secure broker' */
    );
}
exports.isACECardViewReactEnabledOnSecureBrokerKSActivated = isACECardViewReactEnabledOnSecureBrokerKSActivated;
/**
 * @internal
 */
function isSurfaceOnBeforeActionKSActivated() {
    return sp_core_library_1._SPKillSwitch.isActivated('96512177-c963-43cb-b876-87ee06253a2c'
    /* '03/01/2023', 'Enable onBeforeAction API.' */
    );
}
exports.isSurfaceOnBeforeActionKSActivated = isSurfaceOnBeforeActionKSActivated;
/**
 * @internal
 */
function isDisableVivaConnectionsAnalyticsEventsKSActivated() {
    return sp_core_library_1._SPKillSwitch.isActivated('1cc95d27-414e-4c99-9785-d5badd9ec73b'
    /* '07/19/2023', 'Add KS for Disabling of Logging of Viva Connections Events via new logger' */
    );
}
exports.isDisableVivaConnectionsAnalyticsEventsKSActivated = isDisableVivaConnectionsAnalyticsEventsKSActivated;
function isFixOnBeforeActionExternalLinkTelemetryKSActivated() {
    return sp_core_library_1._SPKillSwitch.isActivated('b80ce106-cdb0-42b6-87ab-0b3257902e37'
    /* '03/16/2023', 'Fix onBeforeAction externalLink to openUrl when loggging telemetry.'*/
    );
}
exports.isFixOnBeforeActionExternalLinkTelemetryKSActivated = isFixOnBeforeActionExternalLinkTelemetryKSActivated;
function isIconAltTextNotOverriddenKSActivated() {
    return sp_core_library_1._SPKillSwitch.isActivated('60de4d2f-cafc-48bc-a95f-27b630e60daf'
    /* '03/27/2023', 'Ensure icon alt text is not overridden by undefined or empty string value.' */
    );
}
exports.isIconAltTextNotOverriddenKSActivated = isIconAltTextNotOverriddenKSActivated;
function isRemovingViewNavigatorInstantiationFromConstructorKSActivated() {
    return sp_core_library_1._SPKillSwitch.isActivated('aafb30e0-3b0d-4f43-a2e9-59bc2ef942dc'
    /* '03/29/2023', 'Removing View Navigator instantiation from the constructor' */
    );
}
exports.isRemovingViewNavigatorInstantiationFromConstructorKSActivated = isRemovingViewNavigatorInstantiationFromConstructorKSActivated;
function isAddAppHostTypeToCardActionKSActivated() {
    return sp_core_library_1._SPKillSwitch.isActivated('06dfa773-bc74-41ca-9496-5ebb4afdfd5e'
    /* '04/13/2023', 'Add appHostType to card action engagement logs.' */
    );
}
exports.isAddAppHostTypeToCardActionKSActivated = isAddAppHostTypeToCardActionKSActivated;
function isLogEngagementExtraDataKSActivated() {
    return sp_core_library_1._SPKillSwitch.isActivated('4bd7e0a9-c17a-411b-bda8-1e130c7346ec'
    /* '04/13/2023', 'Check for sending engagement extra data coming from the specific card' */
    );
}
exports.isLogEngagementExtraDataKSActivated = isLogEngagementExtraDataKSActivated;
/**
 * @internal
 */
function isSupportActionExecuteKSActivated() {
    return sp_core_library_1._SPKillSwitch.isActivated('9050e5ae-5e51-4939-9394-7be770d5de93'
    /* '04/20/2023', 'Support Adaptive Card Action.Execute' */
    );
}
exports.isSupportActionExecuteKSActivated = isSupportActionExecuteKSActivated;
function isLockUpdateKSActivated() {
    return sp_core_library_1._SPKillSwitch.isActivated('29970e19-de66-45b7-adcf-74eed779ea77'
    /* '05/11/23', 'Lock update on ACE' */
    );
}
exports.isLockUpdateKSActivated = isLockUpdateKSActivated;
function isAddVetoForACEKSActivated() {
    return sp_core_library_1._SPKillSwitch.isActivated('01795526-5047-424b-ab23-026e44862189'
    /* '05/23/23', 'Add asha veto for ACEs.' */
    );
}
exports.isAddVetoForACEKSActivated = isAddVetoForACEKSActivated;
function isMoveBaseEngagementDataInSeparateFunctionKSActivated() {
    return sp_core_library_1._SPKillSwitch.isActivated('f530e9d6-0952-46f2-9d43-5871b6ce1168'
    /* '06/08/23', 'Move collection of base engagement data in separate function.' */
    );
}
exports.isMoveBaseEngagementDataInSeparateFunctionKSActivated = isMoveBaseEngagementDataInSeparateFunctionKSActivated;
function isThrowErrorWhenButtonActionIsUndefined() {
    return sp_core_library_1._SPKillSwitch.isActivated('ce36fe97-2871-4264-9418-b992195b2dfa'
    /* '06/12/23', 'Throw error when button action is undefined.' */
    );
}
exports.isThrowErrorWhenButtonActionIsUndefined = isThrowErrorWhenButtonActionIsUndefined;
function isFixIframeNameKSActive() {
    return sp_core_library_1._SPKillSwitch.isActivated('1c48a38c-d3a7-4440-9c3e-58a56a92f5be'
    /* '06/15/23', 'Fix iframe name for isolated card' */
    );
}
exports.isFixIframeNameKSActive = isFixIframeNameKSActive;
/**
 * @internal
 */
function isEnableHTMLQuickViewKSActivated() {
    return sp_core_library_1._SPKillSwitch.isActivated('5e86c072-4f36-4fff-8691-3a8b5b4a25aa'
    /* '06/20/2023', 'Enable custom html quick view' */
    );
}
exports.isEnableHTMLQuickViewKSActivated = isEnableHTMLQuickViewKSActivated;
function isMovePropertiesFromDataToCardPropertiesKSActivated() {
    return sp_core_library_1._SPKillSwitch.isActivated('bf5fcd98-4eb8-4768-9c33-6690cc5693e3'
    /* '07/18/2023', 'Move properties from data to card properties' */
    );
}
exports.isMovePropertiesFromDataToCardPropertiesKSActivated = isMovePropertiesFromDataToCardPropertiesKSActivated;
function isRemoveFunctionsForMessagingKSActivated() {
    return sp_core_library_1._SPKillSwitch.isActivated('239e997f-e9d4-4202-ac2f-8d2dfda99e69'
    /* '07/18/2023', 'Remove functions from message payload to avoid serialization issues' */
    );
}
exports.isRemoveFunctionsForMessagingKSActivated = isRemoveFunctionsForMessagingKSActivated;
function isAddParentWindowOriginQSPToIsolatedPageKSActivated() {
    return sp_core_library_1._SPKillSwitch.isActivated('1c24002d-a9fe-462b-b168-b020b9817747'
    /* '07/31/2023', 'Add parent window origin query string parameter to isolated page' */
    );
}
exports.isAddParentWindowOriginQSPToIsolatedPageKSActivated = isAddParentWindowOriginQSPToIsolatedPageKSActivated;
function isCatchNullBotResponseDataKSActivated() {
    return sp_core_library_1._SPKillSwitch.isActivated('20bb1d51-eadf-49ec-8617-c21ae9010ba4'
    /* '06/26/23', 'Catch null responses from bot as expected failure' */
    );
}
exports.isCatchNullBotResponseDataKSActivated = isCatchNullBotResponseDataKSActivated;
function isGenerateComponentIdsKSActivated() {
    return sp_core_library_1._SPKillSwitch.isActivated('d2cb3786-200c-408f-a646-6d091eadc1be'
    /* '08/03/2023', 'Generate component ids' */
    );
}
exports.isGenerateComponentIdsKSActivated = isGenerateComponentIdsKSActivated;
function isProvideCardViewIdKSActivated() {
    return sp_core_library_1._SPKillSwitch.isActivated('f34f1dc6-aefd-4c08-8530-f4ad7e7ad0d4'
    /* '08/02/2023', 'Provide Card View Id' */
    );
}
exports.isProvideCardViewIdKSActivated = isProvideCardViewIdKSActivated;
function isPrepareHTMLQVForMessagingKSActivated() {
    return sp_core_library_1._SPKillSwitch.isActivated('b764e8cc-79df-4487-a967-5ade7b428b61'
    /* '09/01/2023', 'Prepare HTML Quick View for Messaging' */
    );
}
exports.isPrepareHTMLQVForMessagingKSActivated = isPrepareHTMLQVForMessagingKSActivated;
