// @microsoft/gulp-core-build-typescript/lib/index.js
"use strict";
// Copyright (c) Microsoft Corporation. All rights reserved. Licensed under the MIT license.
// See LICENSE in the project root for license information.
Object.defineProperty(exports, "__esModule", { value: true });
exports.apiExtractor = exports.lintCmd = exports.tslintCmd = exports.tscCmd = exports.LintCmdTask = exports.TslintCmdTask = exports.TscCmdTask = void 0;
const TscCmdTask_1 = require("./TscCmdTask");
Object.defineProperty(exports, "TscCmdTask", { enumerable: true, get: function () { return TscCmdTask_1.TscCmdTask; } });
const LintCmdTask_1 = require("./LintCmdTask");
Object.defineProperty(exports, "LintCmdTask", { enumerable: true, get: function () { return LintCmdTask_1.LintCmdTask; } });
const TslintCmdTask_1 = require("./TslintCmdTask");
Object.defineProperty(exports, "TslintCmdTask", { enumerable: true, get: function () { return TslintCmdTask_1.TslintCmdTask; } });
const ApiExtractorTask_1 = require("./ApiExtractorTask");
/** @public */
exports.tscCmd = new TscCmdTask_1.TscCmdTask();
/** @public */
exports.tslintCmd = new TslintCmdTask_1.TslintCmdTask();
/** @public */
exports.lintCmd = new LintCmdTask_1.LintCmdTask();
/** @public */
exports.apiExtractor = new ApiExtractorTask_1.ApiExtractorTask();
