// @microsoft/gulp-core-build-typescript/lib/LintCmdTask.js
"use strict";
// Copyright (c) Microsoft Corporation. All rights reserved. Licensed under the MIT license.
// See LICENSE in the project root for license information.
Object.defineProperty(exports, "__esModule", { value: true });
exports.LintCmdTask = void 0;
const path = require("path");
const node_core_library_1 = require("@rushstack/node-core-library");
const RSCTask_1 = require("./RSCTask");
/**
 * @public
 */
class LintCmdTask extends RSCTask_1.RSCTask {
    constructor() {
        super('lint', {
            displayAsError: false
        });
    }
    loadSchema() {
        return node_core_library_1.JsonFile.load(path.resolve(__dirname, 'schemas', 'lint-cmd.schema.json'));
    }
    executeTask() {
        this.initializeRushStackCompiler();
        const rushStackCompiler = this._rushStackCompiler;
        const lintRunner = new rushStackCompiler.LintRunner({
            displayAsError: this.taskConfig.displayAsError,
            fileError: this.fileError.bind(this),
            fileWarning: this.fileWarning.bind(this)
        }, this.buildFolder, this._terminalProvider);
        return lintRunner.invoke();
    }
}
exports.LintCmdTask = LintCmdTask;
