// @microsoft/sp-loader/lib-commonjs/tiny/common/flights.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.isAPEXSPLoaderRemoveSystemJSEnabled = exports.isAPEXSpfxCustomAppPrincipalClientEnabled = exports.isDurableN1Enabled = exports.isOBOForAllTeamsScenariosEnabled = exports.isMinimalStartFlightEnabled = exports.isSafariAuthPatchEnabled = void 0;
var sp_core_library_1 = require("@microsoft/sp-core-library");
/**
 * SppplatSafariAuthPatch
 */
function isSafariAuthPatchEnabled() {
    return sp_core_library_1._SPFlight.isEnabled(687);
}
exports.isSafariAuthPatchEnabled = isSafariAuthPatchEnabled;
/**
 * SppplatMinimalStart
 */
function isMinimalStartFlightEnabled() {
    return sp_core_library_1._SPFlight.isEnabled(1685);
}
exports.isMinimalStartFlightEnabled = isMinimalStartFlightEnabled;
/**
 * APEXUseOboForAllTeamsScenarios
 */
function isOBOForAllTeamsScenariosEnabled() {
    return sp_core_library_1._SPFlight.isEnabled(60244);
}
exports.isOBOForAllTeamsScenariosEnabled = isOBOForAllTeamsScenariosEnabled;
/**
 * SPServiceWorkerSpfxDurableN1
 */
function isDurableN1Enabled() {
    return sp_core_library_1._SPFlight.isEnabled(1522);
}
exports.isDurableN1Enabled = isDurableN1Enabled;
/**
 * APEXSpfxCustomAppPrincipalClient
 * https://spoflights.azurewebsites.net/Flight/Edit/APEXSpfxCustomAppPrincipalClient
 */
function isAPEXSpfxCustomAppPrincipalClientEnabled() {
    return sp_core_library_1._SPFlight.isEnabled(1807);
}
exports.isAPEXSpfxCustomAppPrincipalClientEnabled = isAPEXSpfxCustomAppPrincipalClientEnabled;
/**
 * APEXSPLoaderRemoveSystemJS
 * https://spoflights.azurewebsites.net/Flight/Edit/APEXSPLoaderRemoveSystemJS
 */
function isAPEXSPLoaderRemoveSystemJSEnabled() {
    return sp_core_library_1._SPFlight.isEnabled(60360);
}
exports.isAPEXSPLoaderRemoveSystemJSEnabled = isAPEXSPLoaderRemoveSystemJSEnabled;
