// @microsoft/sp-loader/lib-commonjs/tiny/common/killSwitches.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.isDoNotApplyCustomSpfxPrincipalToIsolationKSActive = exports.isComponentAsEntryModuleKSActive = exports.isTeamsBrowserHostedKSActive = exports.isBotDrivenACEManifestLoadLogicKSActivated = exports.isGetSessionIdFromQueryParametersKSActivated = exports.isLogLoadPathDependenciesFailuresKSActivated = exports.isDontLogRecursiveRetryKSActivated = exports.isDontFallbackOnScriptFailureKSActivated = exports.overrideWorkloadKSActivated = exports.useOBOInWebForSecureBrokerKSActivated = exports.overrideXMLHttpRequestKSActivated = exports.notUseFirstPartyIsolatedAppIdOnSecurebrokerDomain = exports.traceRequireLoaderErrorKSActivated = exports.isVersionMismatchThrowKSActivated = exports.isVersionHeaderKillSwitchActivated = exports.isRejectOnNotFoundManifestsKillSwitchActivated = void 0;
var sp_core_library_1 = require("@microsoft/sp-core-library");
// THIS KILLSWITCH HAS BEEN PULLED!!!
function isRejectOnNotFoundManifestsKillSwitchActivated() {
    return sp_core_library_1._SPKillSwitch.isActivated('40391ecf-2963-4b4d-b6cf-a08365509a80'
    /* '12/18/20', 'Reject the manifest fetch if we don't return an actual manifest' */
    );
}
exports.isRejectOnNotFoundManifestsKillSwitchActivated = isRejectOnNotFoundManifestsKillSwitchActivated;
function isVersionHeaderKillSwitchActivated() {
    return sp_core_library_1._SPKillSwitch.isActivated('996534b3-7540-4ff2-8864-1a20fce3eba7'
    /* '3/2/22', 'Send odsp-web version to service worker' */
    );
}
exports.isVersionHeaderKillSwitchActivated = isVersionHeaderKillSwitchActivated;
function isVersionMismatchThrowKSActivated() {
    return sp_core_library_1._SPKillSwitch.isActivated('676ca369-31fe-4888-bad3-5ecfc6d828cd' /* '3/9/22', 'Throw when requesting a higher version than a pinned manifest' */);
}
exports.isVersionMismatchThrowKSActivated = isVersionMismatchThrowKSActivated;
function traceRequireLoaderErrorKSActivated() {
    return sp_core_library_1._SPKillSwitch.isActivated('b36bb138-515a-420a-9d65-e09d90cc855f' /* '6/15/22', 'trace loader errors to the console for easier debugging.' */);
}
exports.traceRequireLoaderErrorKSActivated = traceRequireLoaderErrorKSActivated;
function notUseFirstPartyIsolatedAppIdOnSecurebrokerDomain() {
    return sp_core_library_1._SPKillSwitch.isActivated('5910ea14-bd58-4817-90dd-933e0274fe1d' /* '5/8/22', 'Use securebroker app id on securebroker domain' */);
}
exports.notUseFirstPartyIsolatedAppIdOnSecurebrokerDomain = notUseFirstPartyIsolatedAppIdOnSecurebrokerDomain;
function overrideXMLHttpRequestKSActivated() {
    return sp_core_library_1._SPKillSwitch.isActivated('d51a8be3-8ec5-4e35-ada6-cceca79c52f8' /* '8/25/22', 'Override XMLHttpRequest to disable adding traceparent property.' */);
}
exports.overrideXMLHttpRequestKSActivated = overrideXMLHttpRequestKSActivated;
function useOBOInWebForSecureBrokerKSActivated() {
    return sp_core_library_1._SPKillSwitch.isActivated('043feae9-5a56-40ff-b175-6f7f813507ff' /* '10/26/22', 'use OBO in web for secure broker' */);
}
exports.useOBOInWebForSecureBrokerKSActivated = useOBOInWebForSecureBrokerKSActivated;
function overrideWorkloadKSActivated() {
    return sp_core_library_1._SPKillSwitch.isActivated('172a5a30-5ef3-4335-b2b7-d58fb63a2d37' /* '11/11/22', Override telemetry workload' */);
}
exports.overrideWorkloadKSActivated = overrideWorkloadKSActivated;
function isDontFallbackOnScriptFailureKSActivated() {
    return sp_core_library_1._SPKillSwitch.isActivated('e81199cb-6f4e-47d1-8fce-75e5102fe9c2' /* '02/10/23', 'Don't fallback to SystemJS on script failure' */);
}
exports.isDontFallbackOnScriptFailureKSActivated = isDontFallbackOnScriptFailureKSActivated;
function isDontLogRecursiveRetryKSActivated() {
    return sp_core_library_1._SPKillSwitch.isActivated('35caf1d8-29c9-46d8-8dd9-a12e494bc78f' /* '02/10/23', 'Don't log recursive retry' */);
}
exports.isDontLogRecursiveRetryKSActivated = isDontLogRecursiveRetryKSActivated;
function isLogLoadPathDependenciesFailuresKSActivated() {
    return sp_core_library_1._SPKillSwitch.isActivated('e1087c1b-61b7-47eb-92a6-34c34e7f4195' /* '02/13/23', 'Log load path dependencies failures' */);
}
exports.isLogLoadPathDependenciesFailuresKSActivated = isLogLoadPathDependenciesFailuresKSActivated;
function isGetSessionIdFromQueryParametersKSActivated() {
    return sp_core_library_1._SPKillSwitch.isActivated('ef42347e-c163-4c8b-9a66-b3cb90ae4ec5' /* '02/14/23', 'Get sessionId for AadTokenProvider configuration from query parameters' */);
}
exports.isGetSessionIdFromQueryParametersKSActivated = isGetSessionIdFromQueryParametersKSActivated;
function isBotDrivenACEManifestLoadLogicKSActivated() {
    return sp_core_library_1._SPKillSwitch.isActivated('6bc17fec-be41-4460-8f40-b31e293e3d91' /* '05/12/23', 'Update manifest loading logic for Bot-Driven ACEs' */);
}
exports.isBotDrivenACEManifestLoadLogicKSActivated = isBotDrivenACEManifestLoadLogicKSActivated;
function isTeamsBrowserHostedKSActive() {
    return sp_core_library_1._SPKillSwitch.isActivated('733da2bd-f34d-4979-8c7b-df7342291267'
    /* '06/22/2023', 'Set teams hosted container' */
    );
}
exports.isTeamsBrowserHostedKSActive = isTeamsBrowserHostedKSActive;
function isComponentAsEntryModuleKSActive() {
    return sp_core_library_1._SPKillSwitch.isActivated('baf76b43-eb61-4a9e-9a85-1a143f3406e2'
    /* '07/17/2023', 'Disables support for utilizing a component as the entry module for a manifest' */
    );
}
exports.isComponentAsEntryModuleKSActive = isComponentAsEntryModuleKSActive;
function isDoNotApplyCustomSpfxPrincipalToIsolationKSActive() {
    return sp_core_library_1._SPKillSwitch.isActivated('2dcbf2bc-78a8-4f80-9b5d-5c515063fd76'
    /* '07/20/2023', 'Disables applying custom SPFx principal to isolation' */
    );
}
exports.isDoNotApplyCustomSpfxPrincipalToIsolationKSActive = isDoNotApplyCustomSpfxPrincipalToIsolationKSActive;
