// @microsoft/sp-http-base/lib-commonjs/spHttpClient/SPHttpClientBatchConfiguration.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.predefinedConfigurations = void 0;
var tslib_1 = require("tslib");
var SPHttpClientCommonConfiguration_1 = tslib_1.__importDefault(require("./SPHttpClientCommonConfiguration"));
var HttpClient_1 = tslib_1.__importDefault(require("../httpClient/HttpClient"));
/**
 * Configuration for SPHttpClientBatch.
 *
 * @remarks
 * The SPHttpClientBatchConfiguration object provides a set of switches for enabling/disabling
 * various features of the SPHttpClientBatch class.  Normally these switches are set
 * (e.g. when calling SPHttpClientBatch.fetch()) by providing one of the predefined defaults
 * from SPHttpClientBatchConfigurations, however switches can also be changed via the
 * SPHttpClientBatchConfiguration.overrideWith() method.
 *
 * @beta
 */
var SPHttpClientBatchConfiguration = /** @class */ (function (_super) {
    tslib_1.__extends(SPHttpClientBatchConfiguration, _super);
    /**
     * Constructs a new instance of SPHttpClientBatchConfiguration with the specified flags.
     * The default values will be used for any flags that are missing or undefined.
     * If overrideFlags is specified, it takes precedence over flags.
     */
    function SPHttpClientBatchConfiguration(flags, overrideFlags) {
        return _super.call(this, flags, overrideFlags) || this;
    }
    /**
     * @override
     */
    SPHttpClientBatchConfiguration.prototype.overrideWith = function (sourceFlags) {
        return new SPHttpClientBatchConfiguration(this.flags, sourceFlags);
    };
    /**
     * @override
     */
    SPHttpClientBatchConfiguration.prototype.initializeFlags = function () {
        _super.prototype.initializeFlags.call(this);
    };
    return SPHttpClientBatchConfiguration;
}(SPHttpClientCommonConfiguration_1.default));
exports.default = SPHttpClientBatchConfiguration;
exports.predefinedConfigurations = {
    v1: new SPHttpClientBatchConfiguration(HttpClient_1.default.configurations.v1, {
        // ISPHttpClientCommonConfiguration:
        jsonRequest: true,
        jsonResponse: true
        // ISPHttpClientBatchConfiguration:
        // (none)
    })
};
