// @microsoft/spfx-heft-plugins/lib/plugins/packageSolutionPlugin/packageSolution/utils/readSharepointAssets.js
"use strict";
/**
 * Using the configuration, reads a list of Sharepoint assets of one of 3 types
 * (element manifests, element files, and upgrade actions). Returns an object with a list
 * of IFile files to include in the package.
 *
 */
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.readSharepointAssets = void 0;
const path = __importStar(require("path"));
const fs = __importStar(require("fs"));
const node_core_library_1 = require("@rushstack/node-core-library");
/**
 * Reads a special folder containing SharePoint assets which should automatically be integrated into a feature.
 * You can think of this as a slightly more useful version of readCustomFeatures, which allows for custom features
 * to be created, but is more cumbersome when you want to do something simple like include an upgrade action into
 * the SPPKG.
 */
function readSharepointAssets(terminal, assets, featureId) {
    return {
        elementManifests: readAssetFiles(terminal, assets.elementManifests, 'Element Manifest', featureId),
        elementFiles: readAssetFiles(terminal, assets.elementFiles, 'Element File'),
        upgradeActions: readAssetFiles(terminal, assets.upgradeActions, 'Upgrade Action', featureId)
    };
}
exports.readSharepointAssets = readSharepointAssets;
function readAssetFiles(terminal, assetFilePaths, assetFileDescription, subdir) {
    return ((assetFilePaths === null || assetFilePaths === void 0 ? void 0 : assetFilePaths.map((filepath) => {
        let contents;
        const filename = path.basename(filepath);
        try {
            contents = fs.readFileSync(filepath);
        }
        catch (err) {
            throw new Error(`Could not find ${assetFileDescription}: ${filepath}`);
        }
        terminal.writeLine(node_core_library_1.Colors.cyan(`Found ${assetFileDescription}: ${filename}`));
        return {
            filename: subdir ? path.posix.join(subdir, filename) : filename,
            contents
        };
    })) || []);
}
