// @microsoft/sp-adaptive-card-extension-base/lib-commonjs/views/card/parameters/FocusedParameters.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.SearchCardView = void 0;
var CardBarComponent_1 = require("../../../components/CardBarComponent");
var SearchBoxComponent_1 = require("../../../components/SearchBoxComponent");
var SearchFooterComponent_1 = require("../../../components/SearchFooterComponent");
var TextComponent_1 = require("../../../components/TextComponent");
/**
 * Helper method to create a Search Card View.
 * @param configuration - search card view configuration.
 *
 * @public
 */
function SearchCardView(configuration) {
    var cardBar = configuration.cardBar, header = configuration.header, body = configuration.body, footer = configuration.footer;
    return {
        cardViewType: 'search',
        image: undefined,
        cardBar: [(0, CardBarComponent_1._cardViewBarConfigurationToParameters)(cardBar)],
        header: [(0, TextComponent_1._textConfigurationToParameters)(header)],
        body: [(0, SearchBoxComponent_1._searchBoxConfigurationToParameters)(body)],
        footer: [(0, SearchFooterComponent_1._searchFooterConfigurationToParameters)(footer)]
    };
}
exports.SearchCardView = SearchCardView;
