// @rushstack/module-minifier/lib/MessagePortMinifier.js
"use strict";
// Copyright (c) Microsoft Corporation. All rights reserved. Licensed under the MIT license.
// See LICENSE in the project root for license information.
Object.defineProperty(exports, "__esModule", { value: true });
exports.MessagePortMinifier = void 0;
const events_1 = require("events");
/**
 * Minifier implementation that outsources requests to the other side of a MessagePort
 * @public
 */
class MessagePortMinifier {
    constructor(port) {
        this.port = port;
        this._callbacks = new Map();
    }
    /**
     * No-op code transform.
     * @param request - The request to process
     * @param callback - The callback to invoke
     */
    minify(request, callback) {
        const { hash } = request;
        const callbacks = this._callbacks.get(hash);
        if (callbacks) {
            callbacks.push(callback);
            return;
        }
        this._callbacks.set(hash, [callback]);
        this.port.postMessage(request);
    }
    async connect() {
        const configHashPromise = (0, events_1.once)(this.port, 'message');
        this.port.postMessage('initialize');
        const configHash = await configHashPromise;
        const callbacks = this._callbacks;
        function handler(message) {
            if (typeof message === 'object') {
                const callbacksForRequest = callbacks.get(message.hash);
                callbacks.delete(message.hash);
                for (const callback of callbacksForRequest) {
                    callback(message);
                }
            }
        }
        this.port.on('message', handler);
        return {
            configHash,
            disconnect: async () => {
                this.port.off('message', handler);
                this.port.close();
            }
        };
    }
}
exports.MessagePortMinifier = MessagePortMinifier;
