// @microsoft/sp-component-base/lib-commonjs/dataCollector/IWebPartLoadData.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.CacheMissReason = exports.CacheStatus = void 0;
/**
 * @internal
 */
var CacheStatus;
(function (CacheStatus) {
    CacheStatus["True"] = "true";
    CacheStatus["False"] = "false";
    CacheStatus["NotAvailable"] = "N/A";
})(CacheStatus = exports.CacheStatus || (exports.CacheStatus = {}));
/**
 * @internal
 */
var CacheMissReason;
(function (CacheMissReason) {
    /**
     * Cache was available but config requested has changed.
     */
    CacheMissReason["CacheConfigMissmatch"] = "CacheConfigMissmatch";
    /**
     * Cache did not arrvie by the time web part requested it.
     */
    CacheMissReason["LateFlush"] = "LateFlush";
    /**
     * Cache is available but there is no cache entry for the web part.
     */
    CacheMissReason["CachedItemNotFound"] = "CachedItemNotFound";
    /**
     * Cache is available, config is also good but is too stale for use.
     */
    CacheMissReason["CacheExpired"] = "CacheExpired";
    /**
     * Cache miss happened but reason is not specified.
     */
    CacheMissReason["NotSpecified"] = "NotSpecified";
    /**
     * Not Applicable if there was cache hit.
     */
    CacheMissReason["NotApplicable"] = "N/A";
})(CacheMissReason = exports.CacheMissReason || (exports.CacheMissReason = {}));
