// @microsoft/sp-component-base/lib-commonjs/dataCollector/WebPartLoadDataCollector.js
"use strict";
// Copyright (c) Microsoft. All rights reserved.
Object.defineProperty(exports, "__esModule", { value: true });
exports.WebPartLoadDataCollector = void 0;
var sp_telemetry_1 = require("@ms/sp-telemetry");
var IWebPartLoadData_1 = require("./IWebPartLoadData");
var sp_core_library_1 = require("@microsoft/sp-core-library");
var NOT_AVAILABLE = 'N/A';
/**
 * This is used internally to collect web part and adaptive card extension's module load, init, and render performances.
 * This file moved from sp-webpart-base to not additional dependency to sp-adaptive-card-extension-base.
 * @internal
 */
var WebPartLoadDataCollector = /** @class */ (function () {
    function WebPartLoadDataCollector() {
    }
    WebPartLoadDataCollector.collect = function (manifest, webPartTag, pageContext) {
        var getTime = function (key) {
            return (sp_telemetry_1._PerformanceLogger.readComponentBreakdown(webPartTag, key) ||
                sp_telemetry_1._PerformanceLogger.readTempData("".concat(webPartTag, ".").concat(key)));
        };
        var start = getTime('start');
        var moduleLoaded = getTime('modulesLoaded');
        var initialized = getTime('init');
        var end = getTime('end');
        var isSpinnyShown = !!getTime('displaySpinner');
        var isMultiGeo = Boolean(pageContext && pageContext.legacyPageContext && pageContext.legacyPageContext.isMultiGeoTenant);
        var dataProvider = NOT_AVAILABLE;
        var layoutID = NOT_AVAILABLE;
        var webPartCacheHit = IWebPartLoadData_1.CacheStatus.NotAvailable;
        var cacheMissReason = IWebPartLoadData_1.CacheMissReason.NotApplicable;
        // Collect data for first-party web parts only. This ideally should be done outside of spfx.
        // VSO#855246 Refactor WebPartLoadDataCollector to move out first-party only logging logic.
        if (manifest.isInternal) {
            var cacheHit = getTime('CacheHit');
            var cacheMiss = getTime('CacheMiss');
            var prefetechedData = getTime('prefetchedData');
            var isCacheApplicable = Boolean(cacheHit || cacheMiss);
            webPartCacheHit = isCacheApplicable
                ? cacheHit
                    ? IWebPartLoadData_1.CacheStatus.True
                    : IWebPartLoadData_1.CacheStatus.False
                : IWebPartLoadData_1.CacheStatus.NotAvailable;
            cacheMissReason = WebPartLoadDataCollector._getCacheMissReason(cacheHit, getTime);
            var dataProviderTime_1 = sp_telemetry_1._PerformanceLogger.now();
            var layoutKey_1 = sp_telemetry_1._PerformanceDataDimensions.Layout;
            var dataProviderKey_1 = sp_telemetry_1._PerformanceDataDimensions.DataProvider;
            var perfBreakDown_1 = sp_telemetry_1._PerformanceLogger.readFullEUPLBreakDown();
            // WebParts which have different layouts and dataproviders, will have perf marks such as
            // webPartTag.dataProvider: [dataProviderName]
            // webPartTag.Layout: [Layout]
            // The logic below will extract the data provider and layout, and log them for perf measurements.
            Object.keys(perfBreakDown_1).forEach(function (key) {
                if (key.indexOf(webPartTag) > -1) {
                    // We only have 1 layout
                    if (layoutID === NOT_AVAILABLE && key.indexOf(layoutKey_1) > -1) {
                        layoutID = key.split(layoutKey_1)[1];
                        // Focus on primary data provider
                    }
                    else if (key.indexOf(dataProviderKey_1) > -1) {
                        // Pick earliest data provider, in case there are multiple
                        if (dataProvider === NOT_AVAILABLE || perfBreakDown_1[key] < dataProviderTime_1) {
                            dataProvider = key.split(dataProviderKey_1)[1];
                            dataProviderTime_1 = perfBreakDown_1[key];
                        }
                    }
                }
            });
            dataProvider = prefetechedData ? 'Prefetch' : dataProvider;
        }
        var webpartPerfFlights = sp_telemetry_1._PerformanceLogger.getWebpartFlightsAndExperiments(webPartTag);
        var syncRenderTime = this._getSyncRenderTime(layoutID, getTime);
        var retVal = {
            alias: manifest.alias,
            isMultiGeo: isMultiGeo,
            isInternal: manifest.isInternal,
            isSpinnyShown: isSpinnyShown,
            manifestId: manifest.id,
            moduleLoadTime: Math.floor(moduleLoaded - start),
            initTime: Math.floor(initialized - moduleLoaded),
            renderTime: Math.floor(end - initialized),
            syncRenderTime: syncRenderTime,
            scenarioId: sp_telemetry_1._PerformanceLogger.getScenarioId(),
            isFullPage: sp_telemetry_1._PerformanceLogger.fullPageLoad,
            dataProvider: dataProvider,
            layout: layoutID,
            mySiteCacheHit: webPartCacheHit,
            cacheMissReason: cacheMissReason,
            flights: webpartPerfFlights,
            storeAppId: manifest.storeAppId,
            mpnId: manifest.mpnId,
            componentDeveloperName: manifest.componentDeveloperName
        };
        retVal.teamsHosted = sp_core_library_1._BrowserUtilities.isTeamsHosted();
        return retVal;
    };
    WebPartLoadDataCollector._getSyncRenderTime = function (layoutID, getTime) {
        var syncRenderTime = undefined;
        syncRenderTime = Math.floor(getTime('syncRender') - getTime('getDataComplete'));
        var postRenderCompleteTime = getTime('postRenderComplete');
        var loadLayoutEnd = getTime('loadLayoutEnd');
        var syncLayoutRenderComplete = getTime('syncLayoutRenderComplete');
        if (loadLayoutEnd && syncLayoutRenderComplete) {
            // render layout and items inside it
            syncRenderTime = Math.floor(syncRenderTime + (syncLayoutRenderComplete - loadLayoutEnd));
        }
        if (postRenderCompleteTime) {
            syncRenderTime = Math.floor(
            // transposeRow + transposeItems + render + syncPostRender
            syncRenderTime + (postRenderCompleteTime - getTime(sp_telemetry_1._PerformanceDataDimensions.Layout + layoutID)));
        }
        return syncRenderTime;
    };
    WebPartLoadDataCollector._getCacheMissReason = function (cacheHit, getTime) {
        if (cacheHit) {
            return IWebPartLoadData_1.CacheMissReason.NotApplicable;
        }
        var mySiteCacheMissOnConfgMismatch = !!getTime('CacheMissConfigMismatch');
        var mySiteCacheMissOnLateFlush = !!getTime('LateFlush');
        var cachedWebPartNotFound = !!getTime('CachedWebPartNotFound');
        var cacheExpired = !!getTime('CacheExpired');
        var cacheMissReason = IWebPartLoadData_1.CacheMissReason.NotSpecified;
        if (mySiteCacheMissOnConfgMismatch) {
            cacheMissReason = IWebPartLoadData_1.CacheMissReason.CacheConfigMissmatch;
        }
        else if (mySiteCacheMissOnLateFlush) {
            cacheMissReason = IWebPartLoadData_1.CacheMissReason.LateFlush;
        }
        else if (cachedWebPartNotFound) {
            cacheMissReason = IWebPartLoadData_1.CacheMissReason.CachedItemNotFound;
        }
        else if (cacheExpired) {
            cacheMissReason = IWebPartLoadData_1.CacheMissReason.CacheExpired;
        }
        return cacheMissReason;
    };
    return WebPartLoadDataCollector;
}());
exports.WebPartLoadDataCollector = WebPartLoadDataCollector;
