// @microsoft/sp-loader/lib-commonjs/DeveloperTools/DeveloperTools.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var tslib_1 = require("tslib");
var React = tslib_1.__importStar(require("react"));
var ReactDOM = tslib_1.__importStar(require("react-dom"));
var DeveloperToolsTab_1 = tslib_1.__importDefault(require("./Models/DeveloperToolsTab"));
var DeveloperToolsConsole_1 = tslib_1.__importDefault(require("./Components/DeveloperToolsConsole"));
var DeveloperToolsConsoleStore_1 = tslib_1.__importDefault(require("./Stores/DeveloperToolsConsoleStore"));
var sp_core_library_1 = require("@microsoft/sp-core-library");
var FlightInspectorTab_resx_1 = tslib_1.__importDefault(require("./Components/DeveloperModules/FlightInspector/FlightInspectorTab.resx"));
var FlightInspectorTab_1 = require("./Components/DeveloperModules/FlightInspector/FlightInspectorTab");
var ServiceWorkerTab_resx_1 = tslib_1.__importDefault(require("./Components/DeveloperModules/ServiceWorker/ServiceWorkerTab.resx"));
var ServiceWorkerTab_1 = require("./Components/DeveloperModules/ServiceWorker/ServiceWorkerTab");
var DeveloperTools = /** @class */ (function () {
    function DeveloperTools() {
    }
    Object.defineProperty(DeveloperTools, "instance", {
        get: function () {
            if (!this._instance) {
                this._instance = new DeveloperTools();
            }
            return this._instance;
        },
        enumerable: false,
        configurable: true
    });
    DeveloperTools.prototype.initialize = function (serviceScope) {
        var developerToolsDiv = document.createElement('div');
        document.body.appendChild(developerToolsDiv);
        // eslint-disable-next-line @microsoft/spfx/pair-react-dom-render-unmount
        ReactDOM.render(React.createElement(DeveloperToolsConsole_1.default), developerToolsDiv);
        if (sp_core_library_1._SPFlight.isEnabled(466 /* SPFXDebugExperiments */)) {
            // Add flight inspector tab for debug allowed environments so it won't expose to 3rd party developers who don't need this.
            this.registerDeveloperToolsTab({
                title: FlightInspectorTab_resx_1.default.tabTitle,
                iconName: 'Airplane',
                component: FlightInspectorTab_1.FlightInspectorTab,
                reactProps: {
                    serviceScope: serviceScope
                }
            });
        }
        // Service Worker
        if (sp_core_library_1._SPFlight.isEnabled(60892 /* SPServiceWorkerSpfxDevTools */)) {
            this.registerDeveloperToolsTab({
                title: ServiceWorkerTab_resx_1.default.tabTitle,
                iconName: 'Plug',
                component: ServiceWorkerTab_1.ServiceWorkerTab,
                reactProps: {
                    serviceScope: serviceScope
                }
            });
        }
    };
    DeveloperTools.prototype.toggleDeveloperTools = function () {
        DeveloperToolsConsoleStore_1.default.showHideConsole(!DeveloperToolsConsoleStore_1.default.visible);
    };
    DeveloperTools.prototype.showHideDeveloperTools = function (show) {
        DeveloperToolsConsoleStore_1.default.showHideConsole(show);
    };
    DeveloperTools.prototype.registerDeveloperToolsTab = function (developerToolsTab) {
        var tab = new DeveloperToolsTab_1.default(developerToolsTab);
        DeveloperToolsConsoleStore_1.default.registerTab(tab);
    };
    return DeveloperTools;
}());
exports.default = DeveloperTools;
