// @microsoft/sp-image-helper/lib-commonjs/ImageHelper.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ImageHelper = void 0;
var tslib_1 = require("tslib");
var PreviewUtility_1 = require("./PreviewUtility");
var SPResourcePath_1 = require("./SPResourcePath");
var ThumbnailUrlGenerator_1 = tslib_1.__importDefault(require("./ThumbnailUrlGenerator"));
var PUBLIC_CALLER_ID = 'PublicApi';
/**
 * ImageHelper exposes the ability to convert a SharePoint url to an optimized thumbnail url.
 * @public
 */
var ImageHelper = /** @class */ (function () {
    function ImageHelper() {
    }
    /**
     * Convert a url to a file or page on SharePoint into an optimized image url.
     *
     * @param request - IImageHelperRequest for the bundled parameters.
     *
     * @remarks
     * - If width and height are provided, then the resulting image will be scaled to the minimum of the width or height.
     * - The url will be most performant if only a width is provided. If a height is provided, the image will take longer to download due to the scaling.
     * - Only use this url at rendering time. It may contain an expiration token. Continue to use the original file's url for serialization.
     *
     * @returns A full url string for the image of the file. If no optimized url can be created (i.e. an external url or an unsupported file type),
     * the original sourceUrl is returned.
     *
     * @public
     */
    ImageHelper.convertToImageUrl = function (request) {
        var thumbnailRequest = {
            spResource: request.sourceUrl,
            fileType: PreviewUtility_1.PreviewUtility.getFileExtension(new SPResourcePath_1.SPResourcePath(request.sourceUrl)),
            width: request.width,
            height: request.height,
            callerId: PUBLIC_CALLER_ID,
            absoluteUrl: window.location.href,
            version: request.version
        };
        return ThumbnailUrlGenerator_1.default.getThumbnailUrl(thumbnailRequest);
    };
    return ImageHelper;
}());
exports.ImageHelper = ImageHelper;
