// @microsoft/sp-image-helper/lib-commonjs/ThumbnailHealingKillSwitches.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.isAdvancedEditVersioningKSActivated = exports.isOptimizeJpegKSActivated = exports.isOptimizePngKSActivated = exports.isPreferNoRedirectUpdateKSActivated = exports.isUrlGlobalReplaceKSActivated = exports.isQuickLinksThumbnailViewKSActivated = exports.isDocVizUrlRewriteKSActivated = exports.isVROOMThumbnailRewriteKSActivated = exports.isCDNRewriteKSActivated = void 0;
var sp_core_library_1 = require("@microsoft/sp-core-library");
/**
 * Contains predefined kill switches across server and client to disable rewrite to specific
 * service during emergency. These kill switches should not be activated in general but only during
 * emergency situation to mitigate live site for a farm or environment as required.
 * In case a kill switch is activated globally to update guid, ensure to also update the server side guid for same.
 *
 * @internal
 */
/**
 * Do not rewrite URL to CND when true
 */
function isCDNRewriteKSActivated() {
    return sp_core_library_1._SPKillSwitch.isActivated('c2151f9e-ff03-4fa5-8852-10f75e817d4e'
    /* '03/01/2020', 'WEX Disable CDN Rewrite' */
    );
}
exports.isCDNRewriteKSActivated = isCDNRewriteKSActivated;
/**
 * Do not rewrite URL to VROOM Thumbnail when true
 */
function isVROOMThumbnailRewriteKSActivated() {
    return sp_core_library_1._SPKillSwitch.isActivated('318c1916-4594-4509-8ab7-cdb100182007'
    /* '03/01/2020', 'WEX Disable VROOM Thumbnail rewrite' */
    );
}
exports.isVROOMThumbnailRewriteKSActivated = isVROOMThumbnailRewriteKSActivated;
/**
 * Do not rewrite URL to getPreview.ashx when true.
 *
 * @remarks
 * Certain environments like Gallatin supports only DocViz
 */
function isDocVizUrlRewriteKSActivated() {
    return sp_core_library_1._SPKillSwitch.isActivated('67d7e188-b33c-4fc2-8862-a072c6d8afb4'
    /* '03/01/2020', 'WEX Disable DocViz URL Rewrite' */
    );
}
exports.isDocVizUrlRewriteKSActivated = isDocVizUrlRewriteKSActivated;
function isQuickLinksThumbnailViewKSActivated() {
    return sp_core_library_1._SPKillSwitch.isActivated('f1135e46-8b4b-41c8-9af2-008a1596012d'
    /* '6/14/2022', 'FILA fix quicklinks thumbnail not shown consistently for files' */
    );
}
exports.isQuickLinksThumbnailViewKSActivated = isQuickLinksThumbnailViewKSActivated;
function isUrlGlobalReplaceKSActivated() {
    return sp_core_library_1._SPKillSwitch.isActivated('2f268f84-0eee-4f0f-a743-fec73e5e76b2'
    /* '08/24/2022', 'WEX Global replace of forward slash and plus symbol in preview urls' */
    );
}
exports.isUrlGlobalReplaceKSActivated = isUrlGlobalReplaceKSActivated;
function isPreferNoRedirectUpdateKSActivated() {
    return sp_core_library_1._SPKillSwitch.isActivated('04e0cf44-13db-4023-bba2-67aab83c48eb'
    /* '04/07/2023', 'Update VROOM params from preferNoRedirect=true to prefer=noRedirect' */
    );
}
exports.isPreferNoRedirectUpdateKSActivated = isPreferNoRedirectUpdateKSActivated;
function isOptimizePngKSActivated() {
    return sp_core_library_1._SPKillSwitch.isActivated('a7860782-db85-42fc-b1dc-dfb0abaafc97'
    /* '06/27/2023', 'Apply optimization param for PNG files' */
    );
}
exports.isOptimizePngKSActivated = isOptimizePngKSActivated;
function isOptimizeJpegKSActivated() {
    return sp_core_library_1._SPKillSwitch.isActivated('f82624b3-0b57-488e-aa1e-4e164321b4e2'
    /* '06/27/2023', 'Apply optimization param for Jpeg files' */
    );
}
exports.isOptimizeJpegKSActivated = isOptimizeJpegKSActivated;
//cross-project KS
function isAdvancedEditVersioningKSActivated() {
    return sp_core_library_1._SPKillSwitch.isActivated('1f0f59d9-b36f-40ab-a9e4-320f79848087'
    /* '08/03/2023', 'thomashorton- Enabled saving over the same file name in advanced editor and cropper' */
    );
}
exports.isAdvancedEditVersioningKSActivated = isAdvancedEditVersioningKSActivated;
