// @microsoft/sp-property-pane/lib-commonjs/propertyPaneFields/propertyPaneLink/IPropertyPaneLink.js
"use strict";
// Copyright (c) Microsoft. All rights reserved.
Object.defineProperty(exports, "__esModule", { value: true });
exports.PopupWindowPosition = void 0;
/**
 * The position of pop up window.
 *
 * @public
 */
var PopupWindowPosition;
(function (PopupWindowPosition) {
    /**
     * PopupWindowPosition would be located in center of screen.
     */
    PopupWindowPosition[PopupWindowPosition["center"] = 0] = "center";
    /**
     * PopupWindowPosition would be located in right top of screen.
     */
    PopupWindowPosition[PopupWindowPosition["rightTop"] = 1] = "rightTop";
    /**
     * PopupWindowPosition would be located in left top of screen.
     */
    PopupWindowPosition[PopupWindowPosition["leftTop"] = 2] = "leftTop";
    /**
     * PopupWindowPosition would be located in right bottom of screen.
     */
    PopupWindowPosition[PopupWindowPosition["rightBottom"] = 3] = "rightBottom";
    /**
     * PopupWindowPosition would be located in left bottom of screen.
     */
    PopupWindowPosition[PopupWindowPosition["leftBottom"] = 4] = "leftBottom";
})(PopupWindowPosition = exports.PopupWindowPosition || (exports.PopupWindowPosition = {}));
