// @microsoft/sp-adaptive-card-extension-base/lib-commonjs/ac-template/imageL1Template.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.imageL1Template = void 0;
var Constants_1 = require("./common/Constants");
var CustomElements_1 = require("./common/CustomElements");
/**
 * @internal
 *
 * TODO: remove this file and related code after full switch to ACE Components: https://onedrive.visualstudio.com/DefaultCollection/WEX!/_workitems/edit/1690788
 */
exports.imageL1Template = {
    $schema: 'http://adaptivecards.io/schemas/adaptive-card.json',
    type: 'AdaptiveCard',
    version: '1.2',
    body: [
        {
            type: 'Container',
            items: [
                {
                    type: 'Image',
                    url: '${imageUrl}',
                    altText: '${imageAltText}'
                },
                {
                    type: 'ColumnSet',
                    spacing: 'none',
                    columns: [
                        {
                            type: 'Column',
                            items: [
                                {
                                    type: 'Icon',
                                    iconProperty: '${iconProperty}',
                                    altText: '${iconAltText}'
                                }
                            ],
                            width: Constants_1.ICON_COLUMN_WIDTH
                        },
                        {
                            type: 'Column',
                            spacing: 'none',
                            items: [CustomElements_1.TITLE_TEXT],
                            width: 'stretch'
                        }
                    ]
                },
                CustomElements_1.BODY_PRIMARY_TEXT
            ]
        }
    ]
};
