// @microsoft/sp-page-context/lib-commonjs/index.js
"use strict";
/**
 * Page context services for the SharePoint Framework
 *
 * @remarks
 * The page context allows an application and its web parts to share common
 * data about the current page, such as its contents, URL, list, user, permissions,
 * navigation menu, etc.
 *
 * @packagedocumentation
 */
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.SPWeb = exports.SPUser = exports.SPTimeZone = exports.SPSite = exports.SPPermission = exports.SPListItem = exports.SPList = exports.SPField = exports._SPFeatureInfo = exports.SPDayOfWeek = exports.SharePointPageContextDataProvider = exports.PageContext = exports.Page = exports.O365GroupAssociationType = exports.O365GroupAssociation = exports.CultureInfo = exports.AzureActiveDirectoryInfo = void 0;
var AzureActiveDirectoryInfo_1 = require("./AzureActiveDirectoryInfo");
Object.defineProperty(exports, "AzureActiveDirectoryInfo", { enumerable: true, get: function () { return __importDefault(AzureActiveDirectoryInfo_1).default; } });
var CultureInfo_1 = require("./CultureInfo");
Object.defineProperty(exports, "CultureInfo", { enumerable: true, get: function () { return __importDefault(CultureInfo_1).default; } });
var O365GroupAssociation_1 = require("./O365GroupAssociation");
Object.defineProperty(exports, "O365GroupAssociation", { enumerable: true, get: function () { return __importDefault(O365GroupAssociation_1).default; } });
/* IO365GroupAssociation,*/
Object.defineProperty(exports, "O365GroupAssociationType", { enumerable: true, get: function () { return O365GroupAssociation_1.O365GroupAssociationType; } });
var Page_1 = require("./Page");
Object.defineProperty(exports, "Page", { enumerable: true, get: function () { return __importDefault(Page_1).default; } });
var PageContext_1 = require("./PageContext");
Object.defineProperty(exports, "PageContext", { enumerable: true, get: function () { return __importDefault(PageContext_1).default; } });
var SharePointPageContextDataProvider_1 = require("./pageManager/SharePointPageContextDataProvider");
Object.defineProperty(exports, "SharePointPageContextDataProvider", { enumerable: true, get: function () { return __importDefault(SharePointPageContextDataProvider_1).default; } });
var SPDayOfWeek_1 = require("./SPDayOfWeek");
Object.defineProperty(exports, "SPDayOfWeek", { enumerable: true, get: function () { return SPDayOfWeek_1.SPDayOfWeek; } });
var SPFeatureInfo_1 = require("./SPFeatureInfo");
Object.defineProperty(exports, "_SPFeatureInfo", { enumerable: true, get: function () { return __importDefault(SPFeatureInfo_1).default; } });
var SPField_1 = require("./SPField");
Object.defineProperty(exports, "SPField", { enumerable: true, get: function () { return __importDefault(SPField_1).default; } });
var SPList_1 = require("./SPList");
Object.defineProperty(exports, "SPList", { enumerable: true, get: function () { return __importDefault(SPList_1).default; } });
var SPListItem_1 = require("./SPListItem");
Object.defineProperty(exports, "SPListItem", { enumerable: true, get: function () { return __importDefault(SPListItem_1).default; } });
var SPPermission_1 = require("./SPPermission");
Object.defineProperty(exports, "SPPermission", { enumerable: true, get: function () { return __importDefault(SPPermission_1).default; } });
var SPSite_1 = require("./SPSite");
Object.defineProperty(exports, "SPSite", { enumerable: true, get: function () { return __importDefault(SPSite_1).default; } });
var SPTimeZone_1 = require("./SPTimeZone");
Object.defineProperty(exports, "SPTimeZone", { enumerable: true, get: function () { return __importDefault(SPTimeZone_1).default; } });
var SPUser_1 = require("./SPUser");
Object.defineProperty(exports, "SPUser", { enumerable: true, get: function () { return __importDefault(SPUser_1).default; } });
var SPWeb_1 = require("./SPWeb");
Object.defineProperty(exports, "SPWeb", { enumerable: true, get: function () { return __importDefault(SPWeb_1).default; } });
