// @microsoft/sp-page-context/lib-commonjs/O365GroupAssociation.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.O365GroupAssociationType = void 0;
var sp_core_library_1 = require("@microsoft/sp-core-library");
/**
 * {@inheritDoc  O365GroupAssociation."type"}
 * @alpha
 */
var O365GroupAssociationType;
(function (O365GroupAssociationType) {
    O365GroupAssociationType[O365GroupAssociationType["Unknown"] = 0] = "Unknown";
    O365GroupAssociationType[O365GroupAssociationType["Private"] = 1] = "Private";
    O365GroupAssociationType[O365GroupAssociationType["Public"] = 2] = "Public";
})(O365GroupAssociationType = exports.O365GroupAssociationType || (exports.O365GroupAssociationType = {}));
/**
 * This class is primarily used with the `PageContext` class. If the site is associated with an O365
 * group, this object provides contextual information about the group.
 *
 * @alpha
 */
var O365GroupAssociation = /** @class */ (function () {
    /**
     * @internal
     */
    function O365GroupAssociation(data) {
        this.id = sp_core_library_1.Guid.parse(data.id);
        switch (data.type) {
            case 'Private':
                this.type = O365GroupAssociationType.Private;
                break;
            case 'Public':
                this.type = O365GroupAssociationType.Public;
                break;
            default:
                this.type = O365GroupAssociationType.Unknown;
                break;
        }
    }
    return O365GroupAssociation;
}());
exports.default = O365GroupAssociation;
