// @microsoft/sp-page-context/lib-commonjs/SPSite.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var tslib_1 = require("tslib");
var sp_core_library_1 = require("@microsoft/sp-core-library");
var O365GroupAssociation_1 = tslib_1.__importDefault(require("./O365GroupAssociation"));
/**
 * This class is primarily used with the `PageContext` class.  It provides contextual
 * information for the SharePoint site collection ("SPSite") that hosts the page.
 *
 * @public
 */
var SPSite = /** @class */ (function () {
    /**
     * @internal
     */
    function SPSite(data) {
        this.absoluteUrl = data.absoluteUrl;
        this.cdnPrefix = data.cdnPrefix || '';
        this.classification = data.classification;
        this.correlationId = sp_core_library_1.Guid.parse(data.correlationId);
        this.group = data.group ? new O365GroupAssociation_1.default(data.group) : undefined;
        this.id = sp_core_library_1.Guid.parse(data.id);
        this.isNoScriptEnabled = data.isNoScriptEnabled;
        this.recycleBinItemCount = data.recycleBinItemCount;
        this.serverRelativeUrl = data.serverRelativeUrl;
        this.serverRequestPath = data.serverRequestPath;
        this.sitePagesEnabled = data.sitePagesEnabled;
        this.sitePagesFeatureVersion = data.sitePagesFeatureVersion ? data.sitePagesFeatureVersion : 0;
    }
    return SPSite;
}());
exports.default = SPSite;
