// @microsoft/sp-page-context/lib-commonjs/SPTimeZone.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
/**
 * It provides timezone information from SharePoint. A conversion from UTC to local
 * time can be calculated by using the following formula.
 *
 * @remarks
 *
 * ```
 * isDaylightSavingsTime = isCurrentDate after DaylightSavingsDate but before the StandardDate
 * LocalTime = UTCTime - Offset - (isDaylightSavingsTime ? daylightOffset : standardOffset)
 * June 11, 2 A.M. PST = June 11, 9 A.M. - (480 minutes) - (-60  minutes)
 * ```
 *
 * @beta
 */
var SPTimeZone = /** @class */ (function () {
    /**
     * @internal
     */
    function SPTimeZone(data) {
        this.daylightDate = data.daylightDate;
        this.daylightOffset = data.daylightOffset;
        this.description = data.description;
        this.id = data.id;
        this.offset = data.offset;
        this.standardDate = data.standardDate;
        this.standardOffset = data.standardOffset;
    }
    return SPTimeZone;
}());
exports.default = SPTimeZone;
