// @microsoft/sp-page-context/lib-commonjs/SPUser.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var tslib_1 = require("tslib");
var SPTimeZone_1 = tslib_1.__importDefault(require("./SPTimeZone"));
/**
 * This class is primarily used with the `PageContext` class.  It provides contextual information
 * for the SharePoint user that is accessing the page.
 *
 * @public
 */
var SPUser = /** @class */ (function () {
    /**
     * @internal
     */
    function SPUser(data) {
        this.displayName = data.displayName || '';
        this.email = data.email || '';
        this.isAnonymousGuestUser = data.isAnonymousGuestUser;
        this.isExternalGuestUser = data.isExternalGuestUser;
        this.loginName = data.loginName || '';
        // Remove the or statement when changes reach Prod everywhere.
        this.preferUserTimeZone = data.preferUserTimeZone || false;
        this.timeZoneInfo = data.timeZoneInfo ? new SPTimeZone_1.default(data.timeZoneInfo) : undefined;
        this.firstDayOfWeek = data.firstDayOfWeek;
    }
    return SPUser;
}());
exports.default = SPUser;
