// @microsoft/sp-page-context/lib-commonjs/SPWeb.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var tslib_1 = require("tslib");
var sp_core_library_1 = require("@microsoft/sp-core-library");
var SPPermission_1 = tslib_1.__importDefault(require("./SPPermission"));
var SPTimeZone_1 = tslib_1.__importDefault(require("./SPTimeZone"));
/**
 * This class is primarily used with the `PageContext` class. It provides contextual
 * information for the SharePoint site ("SPWeb") that hosts the page.
 *
 * @public
 */
var SPWeb = /** @class */ (function () {
    /**
     * @internal
     */
    function SPWeb(data) {
        this.absoluteUrl = data.absoluteUrl;
        this.id = sp_core_library_1.Guid.parse(data.id);
        this.isAppWeb = data.isAppWeb;
        this.language = data.language;
        this.languageName = data.languageName || '';
        this.logoUrl = data.logoUrl;
        this.permissions = new SPPermission_1.default(data.permissions);
        this.timeZoneInfo = data.timeZoneInfo ? new SPTimeZone_1.default(data.timeZoneInfo) : undefined;
        this.firstDayOfWeek = data.firstDayOfWeek;
        this.serverRelativeUrl = data.serverRelativeUrl;
        this.templateName = data.templateName || '';
        this.title = data.title || '';
        this.description = data.description || '';
    }
    return SPWeb;
}());
exports.default = SPWeb;
