// @microsoft/sp-loader/lib-commonjs/DeveloperTools/Components/DeveloperModules/FlightInspector/FlightInspectorTab.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.FlightInspectorTab = void 0;
var tslib_1 = require("tslib");
var React = tslib_1.__importStar(require("react"));
var FlightInspectorTab_module_scss_1 = tslib_1.__importDefault(require("./FlightInspectorTab.module.scss"));
var SPComponentLoader_1 = require("../../../../loader/SPComponentLoader");
var FlightInspectorTab = /** @class */ (function (_super) {
    tslib_1.__extends(FlightInspectorTab, _super);
    function FlightInspectorTab() {
        var _this = _super !== null && _super.apply(this, arguments) || this;
        _this._rootRef = React.createRef();
        _this._handleCompare = function (theOtherFullString) {
            return _this._flightInspector.compare(theOtherFullString);
        };
        _this._loadFlightInspector = function () {
            void SPComponentLoader_1.SPComponentLoader.loadComponentById('a7f8537f-f742-423f-ad20-ed8feaaf0805' /* sp-flight-inspector */).then(function (module) {
                var instance = module._SPFlightInspector.getInstance(_this.props.serviceScope);
                module._renderFlightViewer(_this._rootRef.current, {
                    enabledFlights: instance.getEnabledFlights(),
                    enabledExperiments: instance.getExperimentVariants(),
                    activatedKillSwitches: instance.getActivatedKillSwitches(),
                    fullString: instance.getFullString(),
                    onCompare: _this._handleCompare,
                    limitMaxSize: false,
                    isPerfQuarantine: instance.getIsPerfQuarantine()
                });
                _this._flightInspector = instance;
            });
        };
        return _this;
    }
    FlightInspectorTab.prototype.componentDidMount = function () {
        this._loadFlightInspector();
    };
    FlightInspectorTab.prototype.render = function () {
        return (React.createElement("div", { className: FlightInspectorTab_module_scss_1.default.root },
            React.createElement("div", { className: FlightInspectorTab_module_scss_1.default.centerAlignedItem, ref: this._rootRef })));
    };
    return FlightInspectorTab;
}(React.Component));
exports.FlightInspectorTab = FlightInspectorTab;
