// @microsoft/sp-component-base/lib-commonjs/test-api/MockComponentContext.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.MockComponentContext = void 0;
var tslib_1 = require("tslib");
var BaseComponentContext_1 = tslib_1.__importDefault(require("../BaseComponentContext"));
var MockComponentContext = /** @class */ (function (_super) {
    tslib_1.__extends(MockComponentContext, _super);
    function MockComponentContext(mockRootScope) {
        return _super.call(this, {
            parentServiceScope: mockRootScope,
            manifest: {
                componentType: 'Application',
                alias: 'alias',
                version: '1.0',
                loaderConfig: {
                    internalModuleBaseUrls: ['mock info'],
                    entryModuleId: 'eId',
                    scriptResources: {}
                },
                manifestVersion: 1,
                id: '79820548-6e95-4fa8-80a0-b366e6725f71'
            },
            loggingTag: 'logtag',
            instanceId: 'iId'
        }) || this;
    }
    return MockComponentContext;
}(BaseComponentContext_1.default));
exports.MockComponentContext = MockComponentContext;
