// @microsoft/sp-module-interfaces/lib-commonjs/manifestSchemas/IClientSideWebPartManifest.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.PredefinedGroup = void 0;
/**
 * Predefined web part group.
 *
 * @beta
 */
var PredefinedGroup;
(function (PredefinedGroup) {
    /**
     * Text, media and content.
     *
     * This group includes web parts that display text, multi-media, documents, information from the web, and other
     * rich content.
     *
     * @beta
     */
    PredefinedGroup["TextMediaAndContent"] = "cf066440-0614-43d6-98ae-0b31cf14c7c3";
    /**
     * Discovery.
     *
     * This group includes web parts that organize, group, and filter content to help users discover information.
     *
     * @beta
     */
    PredefinedGroup["DocumentsListsAndLibraries"] = "1edbd9a8-0bfb-4aa2-9afd-14b8c45dd489";
    /**
     * Communication and collaboration.
     *
     * This group includes web parts that facilitate information sharing, team work, and social interactions.
     *
     * @beta
     */
    PredefinedGroup["Feeds"] = "75e22ed5-fa14-4829-850a-c890608aca2d";
    /**
     * Planning and process.
     *
     * This group includes web parts that empower team productivity with the use of planning and process tools.
     *
     * @beta
     */
    PredefinedGroup["NewsPeopleAndEvents"] = "1bc7927e-4a5e-4520-b540-71305c79c20a";
    /**
     * Business and intelligence.
     *
     * This group includes web parts for tracking and analyzing data, and for integrating business flow with pages.
     *
     * @beta
     */
    PredefinedGroup["DataAnalysis"] = "4aca9e90-eff5-4fa1-bac7-728f5f157b66";
    /**
     * Regional information
     *
     * This group includes web parts that display information based on current region and geographical location
     *
     * @beta
     */
    PredefinedGroup["RegionalInformation"] = "cfc8bda5-cb9b-49e3-8526-2ee6e52b256a";
    /**
     * Other.
     *
     * This group includes web parts not in other groups.
     *
     * @beta
     */
    PredefinedGroup["Advanced"] = "5c03119e-3074-46fd-976b-c60198311f70";
    /**
     * Other.
     *
     * This group includes local web parts.
     *
     * @beta
     */
    PredefinedGroup["Local"] = "8b7bf6f1-a56a-4aa3-8657-7eb6e7e6af61";
})(PredefinedGroup = exports.PredefinedGroup || (exports.PredefinedGroup = {}));
