// @microsoft/sp-property-pane/lib-commonjs/common/KillSwitches.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.isPropertyPaneHostKSActivated = exports.isPropertyPaneGroupNameAndContentA11yKSActivated = exports.isApplyButtonFocusKSActivated = exports.isHideTitleBarKSActivated = exports.isImageChangeButtonFocusedKSActivated = exports.isIncludeChoiceGroupLabelWithRadioButtonAriaLabelKSActivated = exports.isFluentIconPickerPerformanceKSActivated = exports.isActivateActionOnRefreshKSActivated = exports.isLearnMoreAriaLabelKSActivated = exports.isHybridIconFontFamilyActivated = exports.isTopActionsKSActivated = exports.isFixSecondaryGroupKeyboardNavKSActivated = exports.isImageDeleteButtonKSActive = exports.styleChoiceGroupUsingCustomCSSKSActivated = void 0;
var sp_core_library_1 = require("@microsoft/sp-core-library");
var KillSwitches = /** @class */ (function () {
    function KillSwitches() {
    }
    return KillSwitches;
}());
exports.default = KillSwitches;
function styleChoiceGroupUsingCustomCSSKSActivated() {
    return sp_core_library_1._SPKillSwitch.isActivated('bd838357-d9ed-4798-b68c-ea458d17ae91'
    /* '10/18/2021', 'Restore a subset of custom CSS styles for ChoiceGroup removed by KS cb38ec07-5336-4982-8185-8fd7790166d5 and fix line height and text width issue' */
    );
}
exports.styleChoiceGroupUsingCustomCSSKSActivated = styleChoiceGroupUsingCustomCSSKSActivated;
function isImageDeleteButtonKSActive() {
    return sp_core_library_1._SPKillSwitch.isActivated('a8fb6b88-a963-410e-bd83-222b00c7b5da'
    /* '10/29/2021', 'By default, show delete button' */
    );
}
exports.isImageDeleteButtonKSActive = isImageDeleteButtonKSActive;
function isFixSecondaryGroupKeyboardNavKSActivated() {
    return sp_core_library_1._SPKillSwitch.isActivated('e9490b7c-6ea4-471c-8ccc-4a6d7a201829'
    /* '1/4/2022', 'Fix keyboard nav for secondary group option in the property pane' */
    );
}
exports.isFixSecondaryGroupKeyboardNavKSActivated = isFixSecondaryGroupKeyboardNavKSActivated;
function isTopActionsKSActivated() {
    return sp_core_library_1._SPKillSwitch.isActivated('1b316852-63ed-4aa2-a60c-f9608d937195'
    /* '4/01/2022', 'KillSwitch for the Top Actions feature' */
    );
}
exports.isTopActionsKSActivated = isTopActionsKSActivated;
// cross-project KS
function isHybridIconFontFamilyActivated() {
    return sp_core_library_1._SPKillSwitch.isActivated('65e35e80-07f0-4ed0-bb25-2cfdb4ab6bd5'
    /* '06/27/2022', 'Update MDL2 icon font family to the new hybrid font family to fix font not loading' */
    );
}
exports.isHybridIconFontFamilyActivated = isHybridIconFontFamilyActivated;
function isLearnMoreAriaLabelKSActivated() {
    return sp_core_library_1._SPKillSwitch.isActivated('74421037-57be-41fa-a22f-446a18fa63df'
    /* '11/02/2022', 'Adding aria label to the property pane learn more links' */
    );
}
exports.isLearnMoreAriaLabelKSActivated = isLearnMoreAriaLabelKSActivated;
function isActivateActionOnRefreshKSActivated() {
    return sp_core_library_1._SPKillSwitch.isActivated('8a933072-1f53-4ee9-b32f-95aced750f25'
    /* '02/07/2023', 'Refresh Action executed when property pane is not on screen' */
    );
}
exports.isActivateActionOnRefreshKSActivated = isActivateActionOnRefreshKSActivated;
function isFluentIconPickerPerformanceKSActivated() {
    return sp_core_library_1._SPKillSwitch.isActivated('4453188e-9e78-4dbe-b5ad-8dcbc492b797'
    /* '02/07/2023', 'Remove debouncer for performance upgrade on component' */
    );
}
exports.isFluentIconPickerPerformanceKSActivated = isFluentIconPickerPerformanceKSActivated;
function isIncludeChoiceGroupLabelWithRadioButtonAriaLabelKSActivated() {
    return sp_core_library_1._SPKillSwitch.isActivated('4453188e-9e78-4dbe-b5ad-8dcbc492b797'
    /* '04/07/2023', 'Different browsers generate different a11y tree. Disable ChoiceGroup aria-label and include it in radio button aria-label.' */
    );
}
exports.isIncludeChoiceGroupLabelWithRadioButtonAriaLabelKSActivated = isIncludeChoiceGroupLabelWithRadioButtonAriaLabelKSActivated;
function isImageChangeButtonFocusedKSActivated() {
    return sp_core_library_1._SPKillSwitch.isActivated('86e8a66b-80f7-4f52-b085-9a3571dfc29c'
    /* '4/28/2023', 'Set focus on the change button after selecting an image from the file picker' */
    );
}
exports.isImageChangeButtonFocusedKSActivated = isImageChangeButtonFocusedKSActivated;
function isHideTitleBarKSActivated() {
    return sp_core_library_1._SPKillSwitch.isActivated('ebeddd3a-683c-4846-9254-8fc217a57f0d'
    /* '5/16/2023', 'Add an option to hide the title bar' */
    );
}
exports.isHideTitleBarKSActivated = isHideTitleBarKSActivated;
function isApplyButtonFocusKSActivated() {
    return sp_core_library_1._SPKillSwitch.isActivated('4b0e4e25-af2c-47b4-aab0-c6d6736ab370'
    /* '7/10/2023', 'set focus after clicking apply button' */
    );
}
exports.isApplyButtonFocusKSActivated = isApplyButtonFocusKSActivated;
function isPropertyPaneGroupNameAndContentA11yKSActivated() {
    return sp_core_library_1._SPKillSwitch.isActivated('66db35ed-35bd-4ef7-aa7b-7bea2c585abf'
    /* '8/25/2023', 'Add the relationship for the PropertyPane's Group name and content for fixing a11y issue' */
    );
}
exports.isPropertyPaneGroupNameAndContentA11yKSActivated = isPropertyPaneGroupNameAndContentA11yKSActivated;
function isPropertyPaneHostKSActivated() {
    return sp_core_library_1._SPKillSwitch.isActivated('e9890567-59de-4b7a-869d-870b3afe1165'
    /* '9/21/2023', 'Disables the canOpen condition set by the propertyPaneHost property' */
    );
}
exports.isPropertyPaneHostKSActivated = isPropertyPaneHostKSActivated;
