// @microsoft/spfx-heft-plugins/lib/plugins/packageSolutionPlugin/packageSolution/xmlGenerators/createComponentXml.js
"use strict";
/**
 * Creates an package's Element.Xml for a WebPart or ClientSideSolution
 */
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const path = __importStar(require("path"));
const xml_1 = __importDefault(require("xml"));
function createComponentXml(xmlSerializationOptions, options, featureId) {
    const contents = {
        Elements: [
            {
                ...createClientSideComponentElement(options)
            }
        ]
    };
    switch (options.manifest.componentType) {
        case 'WebPart':
        case 'AdaptiveCardExtension':
            contents.Elements.push({
                Module: createModule(featureId, options)
            });
            break;
        case 'Application':
        case 'Extension':
        case 'Library':
            break;
        default:
            throw new Error(`Invalid Component Type: ${options.manifest.componentType}`);
    }
    contents.Elements.push({
        _attr: {
            xmlns: 'http://schemas.microsoft.com/sharepoint/'
        }
    });
    // @todo Note that the options.id is NOT necessarily the feature ID
    return {
        filename: path.posix.join(featureId, `${options.manifest.componentType}_${options.id}.xml`),
        contents: Buffer.from((0, xml_1.default)(contents, xmlSerializationOptions))
    };
}
exports.default = createComponentXml;
/**
 * Creates a ClientSideComponent element.
 */
// eslint-disable-next-line @typescript-eslint/no-explicit-any
function createClientSideComponentElement(options) {
    const attributes = {
        Name: options.name,
        Id: options.id,
        ComponentManifest: JSON.stringify(options.manifest),
        Type: options.manifest.componentType
    };
    if (options.requiresCustomScript) {
        attributes.ReturnIfCustomScriptDisabled = !options.requiresCustomScript;
    }
    return {
        ClientSideComponent: [
            {
                _attr: attributes
            }
        ]
    };
}
/**
 * Creates a Module xml element with all the files.
 */
// eslint-disable-next-line @typescript-eslint/no-explicit-any
function createModule(featureId, component) {
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    const result = [
        {
            _attr: {
                Name: component.name,
                Url: '_catalogs/wp',
                List: '113'
            }
        }
    ];
    return result;
}
