// @microsoft/sp-loader/lib-commonjs/stores/ManifestProvider.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var sp_core_library_1 = require("@microsoft/sp-core-library");
var sp_diagnostics_1 = require("@microsoft/sp-diagnostics");
var sp_http_base_1 = require("@microsoft/sp-http-base");
var sp_page_context_1 = require("@microsoft/sp-page-context");
/**
 * @internal
 */
var ManifestProvider = /** @class */ (function () {
    function ManifestProvider(serviceScope, webAbsoluteUrl) {
        sp_core_library_1.Validate.isNotNullOrUndefined(serviceScope, 'serviceScope');
        this._logSource = sp_diagnostics_1._LogSource.create('ManifestProvider');
        this._webAbsoluteUrl = webAbsoluteUrl;
        this._pageContext = serviceScope.consume(sp_page_context_1.PageContext.serviceKey);
        this._httpClient = serviceScope.consume(sp_http_base_1.SPHttpClient.serviceKey);
    }
    /**
     * Given a component id and version, requests its manifest (and all its dependencies) to SharePoint
     * through a REST API.
     * Returns a promise with all the manifests.
     * @param componentId - Id of the requested component
     * @param version - Optional. Version of the requested component
     */
    ManifestProvider.prototype.tryGetManifest = function (componentId, version) {
        return this.tryGetManifests([{ id: componentId, version: version }]);
    };
    /**
     * Given an array of ids and versions, requests their manifest (and all their dependencies) to SharePoint
     * through a REST API.
     * Returns a promise with all the manifests.
     * @param ids - Array with all the id and version pairs
     */
    ManifestProvider.prototype.tryGetManifests = function (ids) {
        var _this = this;
        var qosMonitor = new sp_diagnostics_1._QosMonitor('ManifestProvider.tryGetManifests');
        var webUrl = (this._pageContext.web && this._pageContext.web.absoluteUrl) || this._webAbsoluteUrl;
        return this._httpClient
            .post(webUrl + ManifestProvider._restApiUrl, sp_http_base_1.SPHttpClient.configurations.v1, {
            body: JSON.stringify({
                components: ids,
                project: sp_core_library_1.Session.clientSideApplicationId
            })
        })
            .then(function (response) {
            if (!response.ok) {
                var error = new Error("GetClientSideComponents failed with HTTP status ".concat(response.status));
                qosMonitor.writeUnexpectedFailure('UnsuccessfulResponse', error, {
                    statusCode: response.status,
                    correlationId: response.correlationId
                });
                throw error;
            }
            qosMonitor.writeSuccess({ correlationId: response.correlationId });
            return response.json().then(_this._extractManifests);
        })
            .catch(function (error) {
            sp_diagnostics_1._TraceLogger.logError(_this._logSource, error);
            qosMonitor.writeUnexpectedFailure(undefined, error);
            throw error;
        });
    };
    ManifestProvider.prototype._extractManifests = function (response) {
        return response.value
            .filter(function (qr) { return qr.Status === 0 && !!qr.Manifest; }) // Remove query results that weren't successful or empty manifests
            .map(function (qr) { return JSON.parse(qr.Manifest); }); // Create the manifest object out of the string
    };
    ManifestProvider._restApiUrl = '/_api/web/GetClientSideComponents';
    return ManifestProvider;
}());
exports.default = ManifestProvider;
