// @microsoft/sp-loader/lib-commonjs/error/SPLoaderError.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.SPLoaderErrorCode = void 0;
var tslib_1 = require("tslib");
var sp_core_library_1 = require("@microsoft/sp-core-library");
var Error_resx_1 = tslib_1.__importDefault(require("./Error.resx"));
var Common_resx_1 = tslib_1.__importDefault(require("../loc/Common.resx"));
/*
 * sp-loader error codes.
 */
var SPLoaderErrorCode;
(function (SPLoaderErrorCode) {
    SPLoaderErrorCode[SPLoaderErrorCode["loadComponentMaxRetriesError"] = 0] = "loadComponentMaxRetriesError";
    SPLoaderErrorCode[SPLoaderErrorCode["loadComponentError"] = 1] = "loadComponentError";
    SPLoaderErrorCode[SPLoaderErrorCode["loadComponentReturnsEmptyError"] = 2] = "loadComponentReturnsEmptyError";
    SPLoaderErrorCode[SPLoaderErrorCode["loadComponentReturnsDefaultEmptyError"] = 3] = "loadComponentReturnsDefaultEmptyError";
    SPLoaderErrorCode[SPLoaderErrorCode["loadComponentDependencyError"] = 4] = "loadComponentDependencyError";
    SPLoaderErrorCode[SPLoaderErrorCode["manifestNotFoundError"] = 5] = "manifestNotFoundError";
    SPLoaderErrorCode[SPLoaderErrorCode["loadPathDependencyBlockedByAnotherDependencyError"] = 6] = "loadPathDependencyBlockedByAnotherDependencyError";
    SPLoaderErrorCode[SPLoaderErrorCode["moduleHasUndeclaredDependencyError"] = 7] = "moduleHasUndeclaredDependencyError";
    SPLoaderErrorCode[SPLoaderErrorCode["loadEntryPointError"] = 8] = "loadEntryPointError";
    SPLoaderErrorCode[SPLoaderErrorCode["loadPathDependencyError"] = 9] = "loadPathDependencyError";
    SPLoaderErrorCode[SPLoaderErrorCode["loadComponentDependencyFailoverPathError"] = 10] = "loadComponentDependencyFailoverPathError";
    SPLoaderErrorCode[SPLoaderErrorCode["loadScriptWithStringError"] = 11] = "loadScriptWithStringError";
    SPLoaderErrorCode[SPLoaderErrorCode["urlStatusLocalhostFileNotFoundError"] = 12] = "urlStatusLocalhostFileNotFoundError";
    SPLoaderErrorCode[SPLoaderErrorCode["urlStatusFileNotFoundError"] = 13] = "urlStatusFileNotFoundError";
    SPLoaderErrorCode[SPLoaderErrorCode["urlStatusForbiddenError"] = 14] = "urlStatusForbiddenError";
    SPLoaderErrorCode[SPLoaderErrorCode["urlStatusClientErrorError"] = 15] = "urlStatusClientErrorError";
    SPLoaderErrorCode[SPLoaderErrorCode["urlStatusServerErrorError"] = 16] = "urlStatusServerErrorError";
    SPLoaderErrorCode[SPLoaderErrorCode["urlStatusLocalhostNetworkErrorError"] = 17] = "urlStatusLocalhostNetworkErrorError";
    SPLoaderErrorCode[SPLoaderErrorCode["urlStatusDocLibNetworkErrorError"] = 18] = "urlStatusDocLibNetworkErrorError";
    SPLoaderErrorCode[SPLoaderErrorCode["urlStatusHttpsNetworkErrorError"] = 19] = "urlStatusHttpsNetworkErrorError";
    SPLoaderErrorCode[SPLoaderErrorCode["urlStatusNetworkErrorError"] = 20] = "urlStatusNetworkErrorError";
    SPLoaderErrorCode[SPLoaderErrorCode["urlStatusUndefinedError"] = 21] = "urlStatusUndefinedError";
    SPLoaderErrorCode[SPLoaderErrorCode["failedToCreateGlobalVariable"] = 22] = "failedToCreateGlobalVariable";
    SPLoaderErrorCode[SPLoaderErrorCode["dependencyLoadError"] = 23] = "dependencyLoadError";
    SPLoaderErrorCode[SPLoaderErrorCode["missingPathDependencyError"] = 24] = "missingPathDependencyError";
})(SPLoaderErrorCode = exports.SPLoaderErrorCode || (exports.SPLoaderErrorCode = {}));
/**
 * sp-loader errors.
 *
 * @internal
 */
var SPLoaderError = /** @class */ (function (_super) {
    tslib_1.__extends(SPLoaderError, _super);
    function SPLoaderError(errorCode, innerError, isExpected) {
        var params = [];
        for (var _i = 3; _i < arguments.length; _i++) {
            params[_i - 3] = arguments[_i];
        }
        var _this = _super.call(this, SPLoaderErrorCode[errorCode], sp_core_library_1.Text.format.apply(sp_core_library_1.Text, tslib_1.__spreadArray([SPLoaderError._errorCodeToMessageMap.get(errorCode)], params, false))) || this;
        _this._isExpected = false;
        // Manually set the prototype, as we can no longer extend built-in classes like Error, Array, Map, etc
        // [https://github.com/Microsoft/TypeScript-wiki/blob/master/Breaking-Changes.md#extending-built-ins-like-error-array-and-map-may-no-longer-work](https://github.com/Microsoft/TypeScript-wiki/blob/master/Breaking-Changes.md#extending-built-ins-like-error-array-and-map-may-no-longer-work)
        //
        // Note: the prototype must also be set on any classes which extend this one
        // eslint-disable-next-line @typescript-eslint/no-explicit-any
        _this.__proto__ = SPLoaderError.prototype;
        _this._loaderErrorCode = errorCode;
        _this.innerError = innerError;
        // If the inner error is expected, the error is expected.
        // Otherwise, look at the flag, or set the default to false.
        if (_this.innerError instanceof SPLoaderError) {
            _this._isExpected = _this.innerError.isExpected;
        }
        _this._isExpected = _this._isExpected || isExpected || false;
        return _this;
    }
    Object.defineProperty(SPLoaderError.prototype, "id", {
        /**
         * Error id.
         */
        get: function () {
            return this._loaderErrorCode;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(SPLoaderError.prototype, "category", {
        /**
         * Error category.
         */
        get: function () {
            return 'SPLoaderError';
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(SPLoaderError.prototype, "isExpected", {
        /**
         * True if the error is an expected error.
         * This can happen when it's surfacing an error that happened in the loader,
         * or there is a transient network problem.
         */
        get: function () {
            return this._isExpected;
        },
        enumerable: false,
        configurable: true
    });
    /**
     * Error id to string map
     */
    SPLoaderError._errorCodeToMessageMap = new Map([
        [SPLoaderErrorCode.loadComponentMaxRetriesError, Common_resx_1.default.loadComponentMaxRetriesError],
        [SPLoaderErrorCode.loadComponentError, Error_resx_1.default.loadComponentError],
        [SPLoaderErrorCode.loadComponentReturnsEmptyError, Error_resx_1.default.loadComponentReturnsEmptyError],
        [
            SPLoaderErrorCode.loadComponentReturnsDefaultEmptyError,
            Error_resx_1.default.loadComponentReturnsDefaultEmptyError
        ],
        [SPLoaderErrorCode.loadComponentDependencyError, Error_resx_1.default.loadComponentDependencyError],
        [SPLoaderErrorCode.manifestNotFoundError, Common_resx_1.default.manifestNotFoundError],
        [
            SPLoaderErrorCode.loadPathDependencyBlockedByAnotherDependencyError,
            Error_resx_1.default.loadPathDependencyBlockedByAnotherDependencyError
        ],
        [SPLoaderErrorCode.moduleHasUndeclaredDependencyError, Error_resx_1.default.moduleHasUndeclaredDependencyError],
        [SPLoaderErrorCode.loadEntryPointError, Error_resx_1.default.loadEntryPointError],
        [SPLoaderErrorCode.loadPathDependencyError, Error_resx_1.default.loadPathDependencyError],
        [
            SPLoaderErrorCode.loadComponentDependencyFailoverPathError,
            Error_resx_1.default.loadComponentDependencyFailoverPathError
        ],
        [SPLoaderErrorCode.loadScriptWithStringError, Error_resx_1.default.loadScriptWithStringError],
        [SPLoaderErrorCode.urlStatusLocalhostFileNotFoundError, Error_resx_1.default.urlStatusLocalhostFileNotFoundError],
        [SPLoaderErrorCode.urlStatusFileNotFoundError, Error_resx_1.default.urlStatusFileNotFoundError],
        [SPLoaderErrorCode.urlStatusForbiddenError, Error_resx_1.default.urlStatusForbiddenError],
        [SPLoaderErrorCode.urlStatusClientErrorError, Error_resx_1.default.urlStatusClientErrorError],
        [SPLoaderErrorCode.urlStatusServerErrorError, Error_resx_1.default.urlStatusServerErrorError],
        [SPLoaderErrorCode.urlStatusLocalhostNetworkErrorError, Error_resx_1.default.urlStatusLocalhostNetworkErrorError],
        // Note that the doclib error shows the standard network error message
        [SPLoaderErrorCode.urlStatusDocLibNetworkErrorError, Error_resx_1.default.urlStatusNetworkErrorError],
        [SPLoaderErrorCode.urlStatusHttpsNetworkErrorError, Error_resx_1.default.urlStatusHttpsNetworkErrorError],
        [SPLoaderErrorCode.urlStatusNetworkErrorError, Error_resx_1.default.urlStatusNetworkErrorError],
        [SPLoaderErrorCode.urlStatusUndefinedError, Error_resx_1.default.urlStatusUndefinedError],
        [SPLoaderErrorCode.failedToCreateGlobalVariable, Error_resx_1.default.failedToCreateGlobalVariableError],
        [SPLoaderErrorCode.dependencyLoadError, Error_resx_1.default.dependencyLoadError],
        [SPLoaderErrorCode.missingPathDependencyError, Error_resx_1.default.missingPathDependencyError]
    ] /* fix this typecast, VSO:397687 */);
    return SPLoaderError;
}(sp_core_library_1.SPError));
exports.default = SPLoaderError;
