// @microsoft/sp-loader/lib-commonjs/DeveloperTools/DataProviders/PerformanceDisplayStateProvider.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.PerformanceDisplayStateProvider = void 0;
var tslib_1 = require("tslib");
var sp_core_library_1 = require("@microsoft/sp-core-library");
var sp_telemetry_1 = require("@ms/sp-telemetry");
var sp_telemetry_2 = require("@ms/sp-telemetry");
var PerformanceDisplay_resx_1 = tslib_1.__importDefault(require("../Components/DeveloperModules/PerformanceDisplay/PerformanceDisplay.resx"));
var PerformanceDisplayStateProvider = /** @class */ (function () {
    function PerformanceDisplayStateProvider() {
    }
    /**
     * This function parses through performance data and puts together some performance data (_IPerfItem) objects for the
     * component to display. If errors are encountered, the _errorMessage variable is set and the function returns an
     * empty array. The component checks for the presence of an error message to determine if something went wrong.
     */
    PerformanceDisplayStateProvider.getState = function () {
        var perfItems = [];
        var toReturn = {
            perfItems: perfItems,
            startTime: 0,
            eupl: 0,
            fullPageLoad: 0,
            speedOfLight: 0,
            timeToInteractive: 0,
            clientBuildVersion: sp_telemetry_2._Telemetry.buildNumber
        };
        // Wrap in a try/catch block to ensure we don't break the page if this throws
        try {
            // @todo VSO:237921 - use the correct typings
            // eslint-disable-next-line @typescript-eslint/no-explicit-any
            var data = sp_telemetry_1._PerformanceLogger.getPerformanceData();
            if (!data || !data.W3cNavigationStart || !data.EUPL) {
                toReturn.errorMessage = PerformanceDisplay_resx_1.default.ErrorAccessingPerfDataErrorMessage;
                console.debug(toReturn.errorMessage);
                return toReturn;
            }
            // We are using performance API so start is always 0 corresponding to W3cNavigationStart
            toReturn.startTime = 0;
            toReturn.eupl = data.EUPL;
            toReturn.spoVersion = data.SharePointServerVersion;
            toReturn.correlationId = data.ServerCorrelationId;
            toReturn.CDNStatus = data.CDNStatus;
            toReturn.fullPageLoad = data.spFullCanvasRendered;
            toReturn.timeToInteractive = data.TTI;
            toReturn.speedOfLight = data.allWebPartsAdded;
            toReturn.firstFlushDuration = data.firstFlushDuration;
            toReturn.serverDuration = data.ServerRequestDuration;
            if (data.postFlush) {
                toReturn.postFlush = Math.round(data.postFlush);
            }
            // Would be really nice if JSON.tryParse existed ...
            var euplBreakdown = void 0;
            try {
                euplBreakdown = JSON.parse(data.EUPLBreakdown);
            }
            catch (error) {
                /* no-op */
            }
            if (!euplBreakdown) {
                toReturn.errorMessage = PerformanceDisplay_resx_1.default.ErrorParsingPercievedLatencyErrorMessage;
                console.debug(toReturn.errorMessage);
                return toReturn;
            }
            var apiCallsData = void 0;
            try {
                apiCallsData = JSON.parse(data.APICalls);
            }
            catch (error) {
                /* no-op */
            }
            if (!apiCallsData) {
                toReturn.errorMessage = PerformanceDisplay_resx_1.default.ErrorParsingApiDataErrorMessage;
                console.debug(toReturn.errorMessage);
                return toReturn;
            }
            if (data.W3cRedirectStart && data.W3cRedirectEnd) {
                perfItems.push({
                    id: 'RedirectResponse',
                    name: PerformanceDisplay_resx_1.default.RedirectResponseLabel,
                    startVal: data.W3cRedirectStart - data.W3cNavigationStart,
                    duration: data.W3cRedirectEnd - data.W3cRedirectStart,
                    breakdown: undefined
                });
            }
            if (data.headStart && data.headEnd && data.bodyStart) {
                perfItems.push({
                    id: 'head',
                    name: PerformanceDisplay_resx_1.default.HeadLabel,
                    startVal: Math.round(data.headStart),
                    duration: Math.round(data.headEnd - data.headStart),
                    breakdown: undefined
                });
                perfItems.push({
                    id: 'serverResponseWait',
                    name: PerformanceDisplay_resx_1.default.ServerResponseWaitLabel,
                    startVal: Math.round(data.headEnd),
                    duration: Math.round(data.bodyStart - data.headEnd),
                    breakdown: undefined
                });
                perfItems.push({
                    id: 'body',
                    name: PerformanceDisplay_resx_1.default.BodyExecutionLabel,
                    startVal: Math.round(data.bodyStart),
                    duration: Math.round(data.bodyEnd - data.bodyStart),
                    breakdown: undefined
                });
            }
            if (euplBreakdown.appStart) {
                perfItems.push({
                    id: 'appStart',
                    name: PerformanceDisplay_resx_1.default.ApplicationInitializationLabel,
                    startVal: Math.round(data.bodyStart),
                    duration: Math.round(data.appStart - data.bodyStart),
                    breakdown: undefined
                });
            }
            if (euplBreakdown.spLoaderStart) {
                perfItems.push({
                    id: 'scriptFetchEval',
                    name: PerformanceDisplay_resx_1.default.ScriptFetchEvalLabel,
                    startVal: Math.round(data.W3cResponseEnd - data.W3cNavigationStart),
                    duration: Math.round(euplBreakdown.spLoaderStart - (data.W3cResponseEnd - data.W3cNavigationStart)),
                    breakdown: undefined
                });
            }
            if (euplBreakdown.spLoaderStart && euplBreakdown.appStart) {
                perfItems.push({
                    id: 'spLoaderStart',
                    name: PerformanceDisplay_resx_1.default.SpLoaderStartLabel,
                    startVal: Math.round(euplBreakdown.spLoaderStart),
                    duration: Math.round(euplBreakdown.appStart - euplBreakdown.spLoaderStart),
                    breakdown: undefined
                });
            }
            /**
             * Check if any web parts made API calls, and store them in a data structure per web part
             */
            var apiCalls = {};
            var endVals = {};
            for (var _i = 0, apiCallsData_1 = apiCallsData; _i < apiCallsData_1.length; _i++) {
                var apiCall = apiCallsData_1[_i];
                if (apiCall.name && apiCall.name.indexOf('WebPart') > -1) {
                    if (!apiCalls[apiCall.name]) {
                        apiCalls[apiCall.name] = [];
                        endVals[apiCall.name] = 0;
                    }
                    apiCalls[apiCall.name].push({
                        startVal: apiCall.startTime,
                        duration: apiCall.duration
                    });
                    // Find the latest ending API call for total duration calculation later on
                    endVals[apiCall.name] = Math.max(apiCall.startTime + apiCall.duration, endVals[apiCall.name]);
                }
            }
            /**
             * Get loading breakdown for each web part. Web parts have their loading times broken down further, so an
             * _IWebPartBreakdown object is put together for each one.
             */
            for (var i = 1; i <= 10; i++) {
                var webPart = data["Control".concat(i, "Id")];
                var renderEnd = data["Control".concat(i, "RenderTime")];
                var renderStart = euplBreakdown["".concat(webPart, ".start")];
                if (!webPart || isNaN(renderStart) || isNaN(renderEnd)) {
                    continue;
                }
                var initTime = euplBreakdown["".concat(webPart, ".init")];
                var moduleLoadedTime = euplBreakdown["".concat(webPart, ".modulesLoaded")];
                var loadingDelayed = euplBreakdown["".concat(webPart, ".loadingDelayed")];
                var inViewportLoading = euplBreakdown["".concat(webPart, ".inViewportLoading")];
                var wpBreakdown = {
                    dataFetch: apiCalls[webPart] || [],
                    render: {
                        startVal: inViewportLoading || initTime || moduleLoadedTime || renderStart,
                        duration: renderEnd - (inViewportLoading || initTime || moduleLoadedTime || renderStart)
                    }
                };
                if (moduleLoadedTime && initTime) {
                    // Get breakdown of loading data for web parts
                    if (!isNaN(moduleLoadedTime)) {
                        wpBreakdown.modulesLoaded = {
                            startVal: renderStart,
                            duration: moduleLoadedTime - renderStart
                        };
                    }
                    if (!isNaN(initTime)) {
                        wpBreakdown.init = {
                            startVal: inViewportLoading || moduleLoadedTime,
                            duration: initTime - (inViewportLoading || moduleLoadedTime)
                        };
                    }
                    if (!isNaN(inViewportLoading) && !isNaN(loadingDelayed)) {
                        wpBreakdown.lazyLoading = {
                            startVal: loadingDelayed,
                            duration: inViewportLoading - loadingDelayed
                        };
                    }
                }
                perfItems.push({
                    id: webPart,
                    name: "".concat(PerformanceDisplayStateProvider._getReadableWebpartName(webPart.replace('Load.', ''))),
                    startVal: renderStart,
                    duration: renderEnd - renderStart,
                    breakdown: wpBreakdown
                });
            }
            return toReturn;
        }
        catch (error) {
            toReturn.errorMessage = sp_core_library_1.Text.format(PerformanceDisplay_resx_1.default.UnknownPerformanceDataErrorMessage, error);
            console.debug(toReturn.errorMessage);
            return toReturn;
        }
    };
    /**
     * Splits the WebPart id to access the alias.
     */
    PerformanceDisplayStateProvider._getReadableWebpartName = function (id) {
        if (id.indexOf('WebPart.') === -1) {
            return id;
        }
        /**
         * idString is in the format of WebPart.manifestID.instanceID
         * Example: WebPart.QuickLinksWebPart.a3deb5fc-2f96-4621-9dfe-985955a33833
         * If there is a failure in getting a readable name from the id
         * (change in web parts manifest) log the error and display the
         * generic name 'web part' so data can still be shown.
         */
        try {
            var readableWebPartName = id.split('.')[1];
            return readableWebPartName;
        }
        catch (error) {
            return id;
        }
    };
    return PerformanceDisplayStateProvider;
}());
exports.PerformanceDisplayStateProvider = PerformanceDisplayStateProvider;
