// @microsoft/sp-core-library/lib-commonjs/index.js
"use strict";
/**
 * SharePoint Framework core libraries.
 *
 * @remarks
 * This package provides a foundation of core classes that ensure a consistent character
 * across all the other SharePoint Framework APIs.  Because this package is a dependency
 * of every other package, its design prioritizes small code size and broad applicability,
 * versus flexibility or richness of features.
 *
 * @packagedocumentation
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports._safeWindowOpen = exports._FluentIconVersionedCdnPath = exports.FormDisplayMode = exports._ServiceKeys = exports._Envoy = exports._BROKER_IN_USE_KEY = exports._BrowserUtilities = exports._DeferredClass = exports.EventArgs = exports.Event = exports._SPEventManager = exports.SPEventArgs = exports.SPEvent = exports.Log = exports.SPError = exports._SPExperiment = exports._SPKillSwitch = exports._SPFlight = exports.EnvironmentType = exports.Environment = exports.Session = exports.ServiceScope = exports.ServiceKey = exports.RandomNumberGenerator = exports.TimeProvider = exports.UrlQueryParameterCollection = exports.ResponseUtilities = exports.UrlUtilities = exports.Validate = exports.Version = exports.Text = exports.Guid = exports._ClientManifestData = exports.DisplayMode = exports._OS = exports._Browser = exports._BrowserDetection = void 0;
var tslib_1 = require("tslib");
var BrowserDetection_1 = require("./BrowserDetection");
Object.defineProperty(exports, "_BrowserDetection", { enumerable: true, get: function () { return tslib_1.__importDefault(BrowserDetection_1).default; } });
Object.defineProperty(exports, "_Browser", { enumerable: true, get: function () { return BrowserDetection_1.Browser; } });
Object.defineProperty(exports, "_OS", { enumerable: true, get: function () { return BrowserDetection_1.OS; } });
var DisplayMode_1 = require("./DisplayMode");
Object.defineProperty(exports, "DisplayMode", { enumerable: true, get: function () { return tslib_1.__importDefault(DisplayMode_1).default; } });
var ClientManifestData_1 = require("./ClientManifestData");
Object.defineProperty(exports, "_ClientManifestData", { enumerable: true, get: function () { return tslib_1.__importDefault(ClientManifestData_1).default; } });
var Guid_1 = require("./Guid");
Object.defineProperty(exports, "Guid", { enumerable: true, get: function () { return tslib_1.__importDefault(Guid_1).default; } });
var Text_1 = require("./Text");
Object.defineProperty(exports, "Text", { enumerable: true, get: function () { return tslib_1.__importDefault(Text_1).default; } });
var Version_1 = require("./Version");
Object.defineProperty(exports, "Version", { enumerable: true, get: function () { return tslib_1.__importDefault(Version_1).default; } });
var Validate_1 = require("./Validate");
Object.defineProperty(exports, "Validate", { enumerable: true, get: function () { return tslib_1.__importDefault(Validate_1).default; } });
var UrlUtilities_1 = require("./url/UrlUtilities");
Object.defineProperty(exports, "UrlUtilities", { enumerable: true, get: function () { return tslib_1.__importDefault(UrlUtilities_1).default; } });
var ResponseUtilities_1 = require("./response/ResponseUtilities");
Object.defineProperty(exports, "ResponseUtilities", { enumerable: true, get: function () { return tslib_1.__importDefault(ResponseUtilities_1).default; } });
var UrlQueryParameterCollection_1 = require("./url/UrlQueryParameterCollection");
Object.defineProperty(exports, "UrlQueryParameterCollection", { enumerable: true, get: function () { return tslib_1.__importDefault(UrlQueryParameterCollection_1).default; } });
var TimeProvider_1 = require("./serviceScope/provider/TimeProvider");
Object.defineProperty(exports, "TimeProvider", { enumerable: true, get: function () { return tslib_1.__importDefault(TimeProvider_1).default; } });
var RandomNumberGenerator_1 = require("./serviceScope/provider/RandomNumberGenerator");
Object.defineProperty(exports, "RandomNumberGenerator", { enumerable: true, get: function () { return tslib_1.__importDefault(RandomNumberGenerator_1).default; } });
var ServiceKey_1 = require("./serviceScope/ServiceKey");
Object.defineProperty(exports, "ServiceKey", { enumerable: true, get: function () { return tslib_1.__importDefault(ServiceKey_1).default; } });
var ServiceScope_1 = require("./serviceScope/ServiceScope");
Object.defineProperty(exports, "ServiceScope", { enumerable: true, get: function () { return tslib_1.__importDefault(ServiceScope_1).default; } });
var Session_1 = require("./Session");
Object.defineProperty(exports, "Session", { enumerable: true, get: function () { return tslib_1.__importDefault(Session_1).default; } });
var Environment_1 = require("./Environment");
Object.defineProperty(exports, "Environment", { enumerable: true, get: function () { return tslib_1.__importDefault(Environment_1).default; } });
Object.defineProperty(exports, "EnvironmentType", { enumerable: true, get: function () { return Environment_1.EnvironmentType; } });
var SPFlight_1 = require("./SPFlight");
Object.defineProperty(exports, "_SPFlight", { enumerable: true, get: function () { return tslib_1.__importDefault(SPFlight_1).default; } });
var SPKillSwitch_1 = require("./SPKillSwitch");
Object.defineProperty(exports, "_SPKillSwitch", { enumerable: true, get: function () { return tslib_1.__importDefault(SPKillSwitch_1).default; } });
var SPExperiment_1 = require("./SPExperiment");
Object.defineProperty(exports, "_SPExperiment", { enumerable: true, get: function () { return SPExperiment_1.SPExperiment; } });
var SPError_1 = require("./SPError");
Object.defineProperty(exports, "SPError", { enumerable: true, get: function () { return tslib_1.__importDefault(SPError_1).default; } });
var Log_1 = require("./log/Log");
Object.defineProperty(exports, "Log", { enumerable: true, get: function () { return tslib_1.__importDefault(Log_1).default; } });
// Events
var SPEvent_1 = require("./events/SPEvent");
Object.defineProperty(exports, "SPEvent", { enumerable: true, get: function () { return tslib_1.__importDefault(SPEvent_1).default; } });
var SPEventArgs_1 = require("./events/SPEventArgs");
Object.defineProperty(exports, "SPEventArgs", { enumerable: true, get: function () { return tslib_1.__importDefault(SPEventArgs_1).default; } });
var SPEventManager_1 = require("./events/SPEventManager");
Object.defineProperty(exports, "_SPEventManager", { enumerable: true, get: function () { return tslib_1.__importDefault(SPEventManager_1).default; } });
// Backwards compatibility for bundles that referenced the originally shipped names
// "Event" and "EventArgs" (which turned out to conflict with the DOM API).
var SPEvent_2 = tslib_1.__importDefault(require("./events/SPEvent"));
var SPEventArgs_2 = tslib_1.__importDefault(require("./events/SPEventArgs"));
/**
 * DO NOT USE THIS DEFINITION.
 * @remarks
 * This is a backwards compatibility shim for legacy third-party bundles that imported the
 * old "Event" class.  It was renamed to solve a naming conflict with the brower DOM "Event".
 * @internal
 * @deprecated Use the SPEvent class instead.
 */
var Event = /** @class */ (function (_super) {
    tslib_1.__extends(Event, _super);
    function Event() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    return Event;
}(SPEvent_2.default));
exports.Event = Event;
/**
 * DO NOT USE THIS DEFINITION.
 * @remarks
 * This is a backwards compatibility shim for legacy third-party bundles that imported the
 * old "EventArgs" class.  It was renamed to solve a naming conflict with the brower DOM "Event".
 * @internal
 * @deprecated Use the SPEvent class instead.
 */
var EventArgs = /** @class */ (function (_super) {
    tslib_1.__extends(EventArgs, _super);
    function EventArgs() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    return EventArgs;
}(SPEventArgs_2.default));
exports.EventArgs = EventArgs;
var DeferredClass_1 = require("./deferredClass/DeferredClass");
Object.defineProperty(exports, "_DeferredClass", { enumerable: true, get: function () { return tslib_1.__importDefault(DeferredClass_1).default; } });
var BrowserUtilities_1 = require("./BrowserUtilities");
Object.defineProperty(exports, "_BrowserUtilities", { enumerable: true, get: function () { return BrowserUtilities_1.BrowserUtilities; } });
var BrowserUtilities_2 = require("./BrowserUtilities");
Object.defineProperty(exports, "_BROKER_IN_USE_KEY", { enumerable: true, get: function () { return BrowserUtilities_2.BROKER_IN_USE_KEY; } });
var Envoy_1 = require("./envoy/Envoy");
Object.defineProperty(exports, "_Envoy", { enumerable: true, get: function () { return Envoy_1.Envoy; } });
var ServiceKeys_1 = require("./serviceScope/serviceKeys/ServiceKeys");
Object.defineProperty(exports, "_ServiceKeys", { enumerable: true, get: function () { return ServiceKeys_1.ServiceKeys; } });
var FormDisplayMode_1 = require("./FormDisplayMode");
Object.defineProperty(exports, "FormDisplayMode", { enumerable: true, get: function () { return tslib_1.__importDefault(FormDisplayMode_1).default; } });
/**
 * @internal
 */
var odsp_core_bundle_1 = require("@ms/odsp-core-bundle");
Object.defineProperty(exports, "_FluentIconVersionedCdnPath", { enumerable: true, get: function () { return odsp_core_bundle_1.PRIMARY_VERSIONED_PATH; } });
var SafeWindowOpen_1 = require("./SafeWindowOpen");
Object.defineProperty(exports, "_safeWindowOpen", { enumerable: true, get: function () { return SafeWindowOpen_1.safeWindowOpen; } });
