// @microsoft/sp-adaptive-card-extension-base/lib-commonjs/components/CardButtonComponent.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports._cardButtonParametersToCardButton = exports._cardButtonToConfiguration = exports._cardButtonConfigurationToParameters = void 0;
var tslib_1 = require("tslib");
/**
 * @internal
 */
function _cardButtonConfigurationToParameters(configuration) {
    return tslib_1.__assign(tslib_1.__assign({}, configuration), { componentName: 'cardButton' });
}
exports._cardButtonConfigurationToParameters = _cardButtonConfigurationToParameters;
/**
 * Helper function to convert ICardButton to ICardButtonConfiguration
 *
 * @internal
 */
function _cardButtonToConfiguration(cardButton) {
    return {
        componentName: 'cardButton',
        title: cardButton.title,
        action: cardButton.action,
        style: cardButton.style === 'destructive' ? undefined : cardButton.style,
        id: cardButton.id
    };
}
exports._cardButtonToConfiguration = _cardButtonToConfiguration;
/**
 * Converts parameters to ICardButton
 *
 * @privateRemarks remove when IDC implements support for ACE Components
 *
 * @internal
 */
function _cardButtonParametersToCardButton(parameters) {
    return {
        title: parameters.title,
        action: parameters.action,
        style: parameters.style,
        id: parameters.id
    };
}
exports._cardButtonParametersToCardButton = _cardButtonParametersToCardButton;
