// @microsoft/rush-stack-compiler-4.7/lib/index.js
"use strict";
// Copyright (c) Microsoft Corporation. All rights reserved. Licensed under the MIT license.
// See LICENSE in the project root for license information.
Object.defineProperty(exports, "__esModule", { value: true });
exports.ApiExtractor = exports.Typescript = exports.ToolPaths = exports.TslintRunner = exports.LintRunner = exports.TypescriptCompiler = exports.StandardBuildFolders = exports.RushStackCompilerBase = exports.ApiExtractorRunner = void 0;
/**
 * This package is used with
 * [\@microsoft/rush-stack](https://www.npmjs.com/package/\@microsoft/rush-stack)
 * to select a TypeScript compiler version.
 *
 * It provides a supported set of versions for the following components:
 * - the TypeScript compiler
 * - [tslint](https://github.com/palantir/tslint#readme)
 * - [API Extractor](https://api-extractor.com/)
 *
 * @packageDocumentation
 */
var ApiExtractorRunner_1 = require("./ApiExtractorRunner");
Object.defineProperty(exports, "ApiExtractorRunner", { enumerable: true, get: function () { return ApiExtractorRunner_1.ApiExtractorRunner; } });
var RushStackCompilerBase_1 = require("./RushStackCompilerBase");
Object.defineProperty(exports, "RushStackCompilerBase", { enumerable: true, get: function () { return RushStackCompilerBase_1.RushStackCompilerBase; } });
var StandardBuildFolders_1 = require("./StandardBuildFolders");
Object.defineProperty(exports, "StandardBuildFolders", { enumerable: true, get: function () { return StandardBuildFolders_1.StandardBuildFolders; } });
var TypescriptCompiler_1 = require("./TypescriptCompiler");
Object.defineProperty(exports, "TypescriptCompiler", { enumerable: true, get: function () { return TypescriptCompiler_1.TypescriptCompiler; } });
var LintRunner_1 = require("./LintRunner");
Object.defineProperty(exports, "LintRunner", { enumerable: true, get: function () { return LintRunner_1.LintRunner; } });
var TslintRunner_1 = require("./TslintRunner");
Object.defineProperty(exports, "TslintRunner", { enumerable: true, get: function () { return TslintRunner_1.TslintRunner; } });
var ToolPaths_1 = require("./ToolPaths");
Object.defineProperty(exports, "ToolPaths", { enumerable: true, get: function () { return ToolPaths_1.ToolPaths; } });
var ToolPackages_1 = require("./ToolPackages");
Object.defineProperty(exports, "Typescript", { enumerable: true, get: function () { return ToolPackages_1.Typescript; } });
Object.defineProperty(exports, "ApiExtractor", { enumerable: true, get: function () { return ToolPackages_1.ApiExtractor; } });
