// @microsoft/sp-module-interfaces/lib-commonjs/manifestSchemas/examples/webpart_2.manifest.js
"use strict";
// eslint-disable-next-line @typescript-eslint/no-explicit-any
const webpartManifest = {
    manifestVersion: 2,
    componentType: 'WebPart',
    id: '00000000-0000-0000-0000-000000000000',
    alias: 'myWebPart',
    version: '0.0.1',
    requiresCustomScript: false,
    loaderConfig: {
        internalModuleBaseUrls: ['https://cdn.net/'],
        entryModuleId: 'main.bundle',
        scriptResources: {
            frameworkA: {
                type: 'component',
                version: '1.2.3',
                id: '00000000-0000-0000-0000-000000000000'
            },
            frameworkB: {
                type: 'component',
                version: '1.2.3',
                id: '00000000-0000-0000-0000-000000000000',
                failoverPath: 'script.js'
            },
            frameworkC: {
                type: 'component',
                version: 'latest',
                id: '00000000-0000-0000-0000-000000000000',
                failoverPath: {
                    path: 'script.js'
                }
            },
            frameworkD: {
                type: 'component',
                version: 'latest',
                id: '00000000-0000-0000-0000-000000000000',
                failoverPath: {
                    path: 'script.js',
                    integrity: 'sha384-oqVuAfXRKap7fdgcCY5uykM6+R9GqQ8K/uxy9rx7HNQlGYl1kPzQho1wx4JwY8wC'
                }
            },
            frameworkE: {
                type: 'component',
                version: '1.2.3',
                id: '00000000-0000-0000-0000-000000000000',
                shouldNotPreload: true
            },
            frameworkF: {
                type: 'component',
                version: '1.2.3',
                id: '00000000-0000-0000-0000-000000000000',
                failoverPath: 'script.js',
                shouldNotPreload: true
            },
            frameworkG: {
                type: 'component',
                version: 'latest',
                id: '00000000-0000-0000-0000-000000000000',
                failoverPath: {
                    path: 'script.js'
                },
                shouldNotPreload: true
            },
            frameworkI: {
                type: 'component',
                version: 'latest',
                id: '00000000-0000-0000-0000-000000000000',
                failoverPath: {
                    path: 'script.js',
                    integrity: 'sha384-oqVuAfXRKap7fdgcCY5uykM6+R9GqQ8K/uxy9rx7HNQlGYl1kPzQho1wx4JwY8wC'
                },
                shouldNotPreload: true
            },
            frameworkJ: {
                type: 'component',
                version: '1.2.3',
                id: '00000000-0000-0000-0000-000000000000',
                shouldNotPreload: false
            },
            frameworkK: {
                type: 'component',
                version: '1.2.3',
                id: '00000000-0000-0000-0000-000000000000',
                failoverPath: 'script.js',
                shouldNotPreload: false
            },
            frameworkL: {
                type: 'component',
                version: 'latest',
                id: '00000000-0000-0000-0000-000000000000',
                failoverPath: {
                    path: 'script.js'
                },
                shouldNotPreload: false
            },
            frameworkM: {
                type: 'component',
                version: 'latest',
                id: '00000000-0000-0000-0000-000000000000',
                failoverPath: {
                    path: 'script.js',
                    integrity: 'sha384-oqVuAfXRKap7fdgcCY5uykM6+R9GqQ8K/uxy9rx7HNQlGYl1kPzQho1wx4JwY8wC'
                },
                shouldNotPreload: false
            },
            'main.bundle': {
                type: 'path',
                path: 'bundle.script.js'
            },
            dependencyA: {
                type: 'path',
                path: {
                    path: 'dependencyA.script.js',
                    integrity: 'sha384-oqVuAfXRKap7fdgcCY5uykM6+R9GqQ8K/uxy9rx7HNQlGYl1kPzQho1wx4JwY8wC'
                }
            },
            dependencyB: {
                type: 'path',
                path: {
                    path: 'dependencyB.script.js'
                }
            },
            dependencyC: {
                type: 'path',
                path: 'dependencyC.script.js'
            },
            dependencyD: {
                type: 'path',
                path: {
                    path: 'dependencyD.script.js',
                    integrity: 'sha384-oqVuAfXRKap7fdgcCY5uykM6+R9GqQ8K/uxy9rx7HNQlGYl1kPzQho1wx4JwY8wC'
                },
                shouldNotPreload: true
            },
            dependencyE: {
                type: 'path',
                path: {
                    path: 'dependencyE.script.js'
                },
                shouldNotPreload: true
            },
            dependencyF: {
                type: 'path',
                path: 'dependencyF.script.js',
                shouldNotPreload: true
            },
            dependencyG: {
                type: 'path',
                path: {
                    path: 'dependencyG.script.js',
                    integrity: 'sha384-oqVuAfXRKap7fdgcCY5uykM6+R9GqQ8K/uxy9rx7HNQlGYl1kPzQho1wx4JwY8wC'
                },
                shouldNotPreload: false
            },
            dependencyH: {
                type: 'path',
                path: {
                    path: 'dependencyH.script.js'
                },
                shouldNotPreload: false
            },
            dependencyI: {
                type: 'path',
                path: 'dependencyI.script.js',
                shouldNotPreload: false
            },
            stringsA: {
                type: 'localizedPath',
                defaultPath: 'stringsA_default.js'
            },
            stringsB: {
                type: 'localizedPath',
                defaultPath: {
                    path: 'stringsB_default.js',
                    integrity: 'sha384-oqVuAfXRKap7fdgcCY5uykM6+R9GqQ8K/uxy9rx7HNQlGYl1kPzQho1wx4JwY8wC'
                }
            },
            stringsC: {
                type: 'localizedPath',
                defaultPath: {
                    path: 'stringsC_default.js'
                }
            },
            stringsD: {
                type: 'localizedPath',
                defaultPath: 'stringsD_default.js',
                paths: {
                    'en-us': 'stringsD_en-us.js',
                    'fr-fr': 'stringsD_fr-fr.js'
                }
            },
            stringsE: {
                type: 'localizedPath',
                defaultPath: 'stringsE_default.js',
                paths: {
                    'en-us': {
                        path: 'stringsE_en-us.js'
                    },
                    'fr-fr': 'stringsE_fr-fr.js'
                }
            },
            stringsF: {
                type: 'localizedPath',
                defaultPath: 'stringsF_default.js',
                paths: {
                    'en-us': {
                        path: 'stringsF_en-us.js',
                        integrity: 'sha384-oqVuAfXRKap7fdgcCY5uykM6+R9GqQ8K/uxy9rx7HNQlGYl1kPzQho1wx4JwY8wC'
                    },
                    'fr-fr': 'stringsF_fr-fr.js'
                }
            },
            stringsG: {
                type: 'localizedPath',
                defaultPath: 'stringsG_default.js',
                shouldNotPreload: true
            },
            stringsH: {
                type: 'localizedPath',
                defaultPath: {
                    path: 'stringsH_default.js',
                    integrity: 'sha384-oqVuAfXRKap7fdgcCY5uykM6+R9GqQ8K/uxy9rx7HNQlGYl1kPzQho1wx4JwY8wC'
                },
                shouldNotPreload: true
            },
            stringsI: {
                type: 'localizedPath',
                defaultPath: {
                    path: 'stringsI_default.js'
                },
                shouldNotPreload: true
            },
            stringsJ: {
                type: 'localizedPath',
                defaultPath: 'stringsJ_default.js',
                paths: {
                    'en-us': 'stringsJ_en-us.js',
                    'fr-fr': 'stringsJ_fr-fr.js'
                },
                shouldNotPreload: true
            },
            stringsK: {
                type: 'localizedPath',
                defaultPath: 'stringsK_default.js',
                paths: {
                    'en-us': {
                        path: 'stringsK_en-us.js'
                    },
                    'fr-fr': 'stringsK_fr-fr.js'
                },
                shouldNotPreload: true
            },
            stringsL: {
                type: 'localizedPath',
                defaultPath: 'stringsL_default.js',
                paths: {
                    'en-us': {
                        path: 'stringsL_en-us.js',
                        integrity: 'sha384-oqVuAfXRKap7fdgcCY5uykM6+R9GqQ8K/uxy9rx7HNQlGYl1kPzQho1wx4JwY8wC'
                    },
                    'fr-fr': 'stringsL_fr-fr.js'
                },
                shouldNotPreload: true
            },
            stringsM: {
                type: 'localizedPath',
                defaultPath: 'stringsM_default.js',
                shouldNotPreload: false
            },
            stringsN: {
                type: 'localizedPath',
                defaultPath: {
                    path: 'stringsN_default.js',
                    integrity: 'sha384-oqVuAfXRKap7fdgcCY5uykM6+R9GqQ8K/uxy9rx7HNQlGYl1kPzQho1wx4JwY8wC'
                },
                shouldNotPreload: false
            },
            stringsO: {
                type: 'localizedPath',
                defaultPath: {
                    path: 'stringsO_default.js'
                },
                shouldNotPreload: false
            },
            stringsP: {
                type: 'localizedPath',
                defaultPath: 'stringsP_default.js',
                paths: {
                    'en-us': 'stringsP_en-us.js',
                    'fr-fr': 'stringsP_fr-fr.js'
                },
                shouldNotPreload: false
            },
            stringsQ: {
                type: 'localizedPath',
                defaultPath: 'stringsQ_default.js',
                paths: {
                    'en-us': {
                        path: 'stringsQ_en-us.js'
                    },
                    'fr-fr': 'stringsQ_fr-fr.js'
                },
                shouldNotPreload: false
            },
            stringsR: {
                type: 'localizedPath',
                defaultPath: 'stringsR_default.js',
                paths: {
                    'en-us': {
                        path: 'stringsR_en-us.js',
                        integrity: 'sha384-oqVuAfXRKap7fdgcCY5uykM6+R9GqQ8K/uxy9rx7HNQlGYl1kPzQho1wx4JwY8wC'
                    },
                    'fr-fr': 'stringsR_fr-fr.js'
                },
                shouldNotPreload: false
            },
            nonAmdA: {
                type: 'path',
                path: 'nonAmdA.js',
                globalName: 'nonAmdA'
            },
            nonAmdB: {
                type: 'path',
                path: 'nonAmdB.js',
                globalName: 'nonAmdB',
                globalDependencies: ['nonAmdB']
            }
        }
    },
    preconfiguredEntries: [
        {
            title: { default: 'asdf' },
            description: { default: 'asdf jkl' },
            iconImageUrl: 'https://contoso.com/chart.png',
            properties: {},
            groupId: '1bc7927e-4a5e-4520-b540-71305c79c20a' // Planning and Process
        },
        {
            title: { default: 'asdf', 'en-us': 'foobar' },
            description: { default: 'asdf jkl', 'en-us': 'asdf' },
            iconImageUrl: 'https://contoso.com/chart.png',
            properties: {
                a: 'b',
                c: 'd'
            },
            group: { default: 'asdf' },
            groupId: '0314dcd9-9acb-4ab2-886e-f8881ffb5192' // Customized Group
        },
        {
            title: { default: 'asdf', 'en-us': 'foobar' },
            description: { default: 'asdf jkl', 'en-us': 'asdf' },
            iconImageUrl: 'https://contoso.com/chart.png',
            properties: {
                a: 'b',
                c: 'd'
            },
            group: { default: 'asdf', 'en-us': 'asdfasdf' },
            groupId: '1db70b09-46dc-4fa1-9c97-dcce28dad5d9' // Customized Group
        }
    ]
};
module.exports = webpartManifest;
