// @microsoft/sp-dynamic-data/lib-commonjs/DynamicDataManager.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var tslib_1 = require("tslib");
var sp_core_library_1 = require("@microsoft/sp-core-library");
var sp_diagnostics_1 = require("@microsoft/sp-diagnostics");
var Strings_resx_1 = tslib_1.__importDefault(require("./loc/Strings.resx"));
var DynamicDataEventConstants_1 = require("./common/DynamicDataEventConstants");
var DynamicDataEventNames_1 = require("./common/DynamicDataEventNames");
var QOS_MONITOR_PREFIX = 'DynamicData.DynamicDataManager';
var LOG_SOURCE = sp_diagnostics_1._LogSource.create('DynamicDataManager');
/**
 * Manager for Dynamic Data.
 * This holds a reference to all Dynamic Data Sources and uses the SPEventManager internally to handle
 * notifications for source updates.
 *
 * @remarks
 * Data sources can notify for changes within the data source or for a specific property.
 * Data consumers can register to both changes in a Dynamic Data source and a specific property within the source.
 * There are 3 events to handle this variety of situations:
 *
 * Notifying a change for a property triggers events for the specific property and for "any" property
 * Notifying a change for the whole source triggers events for "any" property and "all" properties.
 *
 * Registering for changes in a property register to events for the specific property and "all" properties.
 * Registering for changes in the whole source register to events for "any" properties.
 *
 * This ensures that regardless of how sources and consumers are configured, events will trigger once and only once for
 * each update within the data source.
 *
 * @internal
 */
var DynamicDataManager = /** @class */ (function () {
    function DynamicDataManager(serviceScope) {
        this._sources = new Map();
        this._sourcesChangedEvent = new sp_core_library_1.SPEvent(DynamicDataEventConstants_1.SOURCES_CHANGED_EVENT_NAME);
    }
    Object.defineProperty(DynamicDataManager.prototype, "sourcesChangedEvent", {
        /**
         * Event that gets raised when the list of Dynamic Data Sources gets updated.
         * @eventproperty
         */
        get: function () {
            return this._sourcesChangedEvent;
        },
        enumerable: false,
        configurable: true
    });
    /**
     * Raises an event to all listeners when a Dynamic Data Source has been updated.
     * @param sourceId - Id of the Dynamic Data Source that is being updated.
     */
    DynamicDataManager.prototype.notifySourceChanged = function (sourceId) {
        sp_core_library_1.Validate.isNonemptyString(sourceId, 'sourceId');
        sp_core_library_1._SPEventManager.instance.raiseEvent(DynamicDataEventNames_1.DynamicDataEventNames.getAllPropertiesChangedEventName(sourceId), {});
        sp_core_library_1._SPEventManager.instance.raiseEvent(DynamicDataEventNames_1.DynamicDataEventNames.getAnyPropertyChangedEventName(sourceId), {});
    };
    /**
     * Raises an event to all listeners when a property in a Dynamic Data Source has been updated.
     * @param sourceId - Id of the Dynamic Data Source whose property is being updated.
     * @param propertyId - Id of the property that is being updated.
     */
    DynamicDataManager.prototype.notifyPropertyChanged = function (sourceId, propertyId) {
        sp_core_library_1.Validate.isNonemptyString(sourceId, 'sourceId');
        sp_core_library_1.Validate.isNonemptyString(propertyId, 'propertyId');
        sp_core_library_1._SPEventManager.instance.raiseEvent(DynamicDataEventNames_1.DynamicDataEventNames.getPropertyChangedEventName(sourceId, propertyId), {});
        sp_core_library_1._SPEventManager.instance.raiseEvent(DynamicDataEventNames_1.DynamicDataEventNames.getAnyPropertyChangedEventName(sourceId), {});
    };
    /**
     * Registers a listener for updates on a Dynamic Data Source.
     */
    DynamicDataManager.prototype.registerSourceChanged = function (sourceId, observer, callback) {
        var extraData = this._createQosExtraData(observer.manifest);
        var qosMonitor = new sp_diagnostics_1._QosMonitor(QOS_MONITOR_PREFIX + '.registerSourceChanged');
        try {
            sp_core_library_1.Validate.isNonemptyString(sourceId, 'sourceId');
            sp_core_library_1.Validate.isNotNullOrUndefined(observer, 'observer');
            sp_core_library_1.Validate.isNotNullOrUndefined(callback, 'callback');
            if (!this._sources.has(sourceId)) {
                throw new Error(sp_core_library_1.Text.format(Strings_resx_1.default.dynamicDataManagerSourceDoesntExist, sourceId));
            }
            // Registering to all properties changed event.
            sp_core_library_1._SPEventManager.instance.registerEvent(DynamicDataEventNames_1.DynamicDataEventNames.getAllPropertiesChangedEventName(sourceId), observer, callback);
            // Registering to any property changed event, as it means that the source is changed.
            sp_core_library_1._SPEventManager.instance.registerEvent(DynamicDataEventNames_1.DynamicDataEventNames.getAnyPropertyChangedEventName(sourceId), observer, callback);
            qosMonitor.writeSuccess(extraData);
        }
        catch (e) {
            qosMonitor.writeUnexpectedFailure('UnhandledRegisterSourceChange', e, extraData);
            throw e;
        }
    };
    /**
     * Unregisters a listener for updates on a Dynamic Data Source.
     */
    DynamicDataManager.prototype.unregisterSourceChanged = function (sourceId, observer, callback) {
        var extraData = this._createQosExtraData(observer.manifest);
        var qosMonitor = new sp_diagnostics_1._QosMonitor(QOS_MONITOR_PREFIX + '.unregisterSourceChanged');
        try {
            sp_core_library_1.Validate.isNonemptyString(sourceId, 'sourceId');
            sp_core_library_1.Validate.isNotNullOrUndefined(observer, 'observer');
            sp_core_library_1.Validate.isNotNullOrUndefined(callback, 'callback');
            if (!this._sources.has(sourceId)) {
                throw new Error(sp_core_library_1.Text.format(Strings_resx_1.default.dynamicDataManagerSourceDoesntExist, sourceId));
            }
            sp_core_library_1._SPEventManager.instance.unregisterEvent(DynamicDataEventNames_1.DynamicDataEventNames.getAllPropertiesChangedEventName(sourceId), observer, callback);
            sp_core_library_1._SPEventManager.instance.unregisterEvent(DynamicDataEventNames_1.DynamicDataEventNames.getAnyPropertyChangedEventName(sourceId), observer, callback);
            qosMonitor.writeSuccess(extraData);
        }
        catch (e) {
            qosMonitor.writeUnexpectedFailure('UnhandledUnregisterSourceChange', e, extraData);
            throw e;
        }
    };
    /**
     * Registers a listener for updates on a Dynamic Data Source.
     */
    DynamicDataManager.prototype.registerPropertyChanged = function (sourceId, propertyId, observer, callback) {
        var extraData = this._createQosExtraData(observer.manifest);
        var qosMonitor = new sp_diagnostics_1._QosMonitor(QOS_MONITOR_PREFIX + '.registerPropertyChanged');
        try {
            sp_core_library_1.Validate.isNonemptyString(sourceId, 'sourceId');
            sp_core_library_1.Validate.isNonemptyString(propertyId, 'propertyId');
            sp_core_library_1.Validate.isNotNullOrUndefined(observer, 'observer');
            sp_core_library_1.Validate.isNotNullOrUndefined(callback, 'callback');
            if (!this._sources.has(sourceId)) {
                throw new Error(sp_core_library_1.Text.format(Strings_resx_1.default.dynamicDataManagerSourceDoesntExist, sourceId));
            }
            // Registering to the specific property
            sp_core_library_1._SPEventManager.instance.registerEvent(DynamicDataEventNames_1.DynamicDataEventNames.getPropertyChangedEventName(sourceId, propertyId), observer, callback);
            // Registering to all properties changed event, as the specific property will also be changed.
            sp_core_library_1._SPEventManager.instance.registerEvent(DynamicDataEventNames_1.DynamicDataEventNames.getAllPropertiesChangedEventName(sourceId), observer, callback);
            qosMonitor.writeSuccess(extraData);
        }
        catch (e) {
            qosMonitor.writeUnexpectedFailure('UnhandledRegisterPropertyChange', e, extraData);
            throw e;
        }
    };
    /**
     * Unregisters a listener for updates on a Dynamic Data Source.
     */
    DynamicDataManager.prototype.unregisterPropertyChanged = function (sourceId, propertyId, observer, callback) {
        var extraData = this._createQosExtraData(observer.manifest);
        var qosMonitor = new sp_diagnostics_1._QosMonitor(QOS_MONITOR_PREFIX + '.unregisterPropertyChanged');
        try {
            sp_core_library_1.Validate.isNonemptyString(sourceId, 'sourceId');
            sp_core_library_1.Validate.isNonemptyString(propertyId, 'propertyId');
            sp_core_library_1.Validate.isNotNullOrUndefined(observer, 'observer');
            sp_core_library_1.Validate.isNotNullOrUndefined(callback, 'callback');
            if (!this._sources.has(sourceId)) {
                throw new Error(sp_core_library_1.Text.format(Strings_resx_1.default.dynamicDataManagerSourceDoesntExist, sourceId));
            }
            sp_core_library_1._SPEventManager.instance.unregisterEvent(DynamicDataEventNames_1.DynamicDataEventNames.getPropertyChangedEventName(sourceId, propertyId), observer, callback);
            sp_core_library_1._SPEventManager.instance.unregisterEvent(DynamicDataEventNames_1.DynamicDataEventNames.getAllPropertiesChangedEventName(sourceId), observer, callback);
            qosMonitor.writeSuccess(extraData);
        }
        catch (e) {
            qosMonitor.writeUnexpectedFailure('UnhandledUnregisterPropertyChange', e, extraData);
            throw e;
        }
    };
    /**
     * Returns a read-only array with all the existing Dynamic Data Sources.
     */
    DynamicDataManager.prototype.getSources = function () {
        var sources = [];
        this._sources.forEach(function (source) { return sources.push(source); });
        return sources;
    };
    /**
     * Returns a Dynamic Data Source based on its id.
     * Returns undefined if the source doesn't exist.
     * @param sourceId - Id of the Dynamic Data Source.
     */
    DynamicDataManager.prototype.tryGetSource = function (sourceId) {
        sp_core_library_1.Validate.isNotNullOrUndefined(sourceId, 'sourceId');
        return this._sources.get(sourceId);
    };
    /**
     * Adds a new Dynamic Data Source to be managed.
     * Throws an error if the source can't be added.
     *
     * @param source - Dynamic Data Source to add.
     */
    DynamicDataManager.prototype.addSource = function (source) {
        this._validateSource(source);
        if (this._sources.has(source.id)) {
            sp_diagnostics_1._TraceLogger.logVerbose(LOG_SOURCE, sp_core_library_1.Text.format(Strings_resx_1.default.dynamicDataManagerSourceAlreadyExists, source.id));
        }
        this._sources.set(source.id, source);
        this._raiseSourcesChangedEvent();
    };
    /**
     * Removes an existing Dynamic Data Source from the manager.
     * @param sourceId - Id of the Dynamic Data Source.
     */
    DynamicDataManager.prototype.removeSource = function (sourceId) {
        var qosMonitor = new sp_diagnostics_1._QosMonitor(QOS_MONITOR_PREFIX + '.removeSource');
        try {
            sp_core_library_1.Validate.isNonemptyString(sourceId, 'sourceId');
            if (this._sources.has(sourceId)) {
                sp_core_library_1._SPEventManager.instance.removeEvent(DynamicDataEventNames_1.DynamicDataEventNames.getAllPropertiesChangedEventName(sourceId));
                sp_core_library_1._SPEventManager.instance.removeEvent(DynamicDataEventNames_1.DynamicDataEventNames.getAnyPropertyChangedEventName(sourceId));
                sp_core_library_1._SPEventManager.instance.removeEventsByPrefix(DynamicDataEventNames_1.DynamicDataEventNames.getPropertyChangedEventPrefix(sourceId));
                this._sources.delete(sourceId);
                this._raiseSourcesChangedEvent();
            }
            qosMonitor.writeSuccess();
        }
        catch (e) {
            qosMonitor.writeUnexpectedFailure(e);
            throw e;
        }
    };
    /**
     * Creates a new extra data object _IQosExtraData.
     */
    DynamicDataManager.prototype._createQosExtraData = function (manifest) {
        if (!manifest) {
            return undefined;
        }
        var qosExtraData = {
            alias: manifest.alias,
            isInternal: manifest.isInternal,
            manifestId: manifest.id
        };
        return qosExtraData;
    };
    DynamicDataManager.prototype._validateSource = function (source) {
        sp_core_library_1.Validate.isNotNullOrUndefined(source, 'source');
        if (source.id.indexOf(DynamicDataEventConstants_1.EVENT_NAME_SEPARATOR) > -1) {
            throw new Error("Source id contains invalid characters, like \"".concat(DynamicDataEventConstants_1.EVENT_NAME_SEPARATOR, "\". Id: \"").concat(source.id, "\"."));
        }
        var regex = /^[a-zA-Z0-9\-_]+$/;
        source.getPropertyDefinitions().forEach(function (def) {
            if (!regex.test(def.id)) {
                throw new Error("Source contains invalid property \"".concat(def.id, "\"."));
            }
        });
    };
    /**
     * Raises an event when the Dynamic Data Sources gets updated.
     * @remarks
     * The event is sticky because sources can be updated before there is anyone listening. This way all clients
     * will get notified that sources have been updated. Further updates are notified in real-time.
     */
    DynamicDataManager.prototype._raiseSourcesChangedEvent = function () {
        sp_core_library_1._SPEventManager.instance.raiseStickyEvent(DynamicDataEventConstants_1.SOURCES_CHANGED_EVENT_NAME, {});
    };
    return DynamicDataManager;
}());
exports.default = DynamicDataManager;
