// @microsoft/sp-dynamic-data/lib-commonjs/common/DynamicDataEventConstants.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.NATIVE_SOURCE_API_EVENT = exports.DYNAMICDATAMANAGER_EXCEPTION = exports.GET_ALLOWED_EVENTS = exports.GET_PROPERTY_DEFINITIONS = exports.GET_ANNOTATED_PROPERTY_VALUE = exports.GET_PROPERTY_VALUE = exports.EVENT_NAME_SEPARATOR = exports.SOURCES_RECEIVED = exports.SOURCES_CHANGED_EVENT_NAME = exports.ALL_PROPERTIES_EVENT_PREFIX = exports.PROPERTY_EVENT_PREFIX = exports.ANY_PROPERTY_EVENT_PREFIX = void 0;
exports.ANY_PROPERTY_EVENT_PREFIX = 'DynamicData_AnyPropertyChanged';
exports.PROPERTY_EVENT_PREFIX = 'DynamicData_PropertyChanged';
exports.ALL_PROPERTIES_EVENT_PREFIX = 'DynamicData_AllPropertiesChanged';
exports.SOURCES_CHANGED_EVENT_NAME = 'DynamicData_SourcesChanged';
exports.SOURCES_RECEIVED = 'DynamicData_All_Sources_Received';
exports.EVENT_NAME_SEPARATOR = '_';
exports.GET_PROPERTY_VALUE = 'DynamicData_PropertyValue';
exports.GET_ANNOTATED_PROPERTY_VALUE = 'DynamicData_Annotated_PropertyValue';
exports.GET_PROPERTY_DEFINITIONS = 'DynamicData_PropertyDefinitions';
exports.GET_ALLOWED_EVENTS = 'DynamicData_Allowed_Events';
exports.DYNAMICDATAMANAGER_EXCEPTION = 'DynamicDataManager_Exception';
exports.NATIVE_SOURCE_API_EVENT = 'Native_Source_Api_Event';
