// @rushstack/localization-utilities/lib/index.js
"use strict";
// Copyright (c) Microsoft Corporation. All rights reserved. Licensed under the MIT license.
// See LICENSE in the project root for license information.
Object.defineProperty(exports, "__esModule", { value: true });
exports.getPseudolocalizer = exports.TypingsGenerator = exports.parseLocFile = exports.parseResx = exports.parseResJson = exports.parseLocJson = void 0;
var parseLocJson_1 = require("./parsers/parseLocJson");
Object.defineProperty(exports, "parseLocJson", { enumerable: true, get: function () { return parseLocJson_1.parseLocJson; } });
var parseResJson_1 = require("./parsers/parseResJson");
Object.defineProperty(exports, "parseResJson", { enumerable: true, get: function () { return parseResJson_1.parseResJson; } });
var parseResx_1 = require("./parsers/parseResx");
Object.defineProperty(exports, "parseResx", { enumerable: true, get: function () { return parseResx_1.parseResx; } });
var LocFileParser_1 = require("./LocFileParser");
Object.defineProperty(exports, "parseLocFile", { enumerable: true, get: function () { return LocFileParser_1.parseLocFile; } });
var TypingsGenerator_1 = require("./TypingsGenerator");
Object.defineProperty(exports, "TypingsGenerator", { enumerable: true, get: function () { return TypingsGenerator_1.TypingsGenerator; } });
var Pseudolocalization_1 = require("./Pseudolocalization");
Object.defineProperty(exports, "getPseudolocalizer", { enumerable: true, get: function () { return Pseudolocalization_1.getPseudolocalizer; } });
